{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Challenge Problem 3: Genome Wide Association Study"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Background:\n",
    "\n",
    "(Adapted From BMI 704, Dr. Chirag Patel)\n",
    "\n",
    "Genome-wide association studies (GWASs) are arguably the most significant biomedical advance in the last decade. Utilizing an observational case-control design, populations with disease are analytically compared to those without to discern patterns in differences in genetic variant, or single nucleotide polymorphism (SNP),frequencies in each of the respective populations. If statistically differences in SNP frequencies in cases vs\n",
    "controls are found, one can conclude the SNP locus may have a causal relationship, implicating a gene, genetic region, or gene regulator in the disease of interest.\n",
    "\n",
    "Analyzing GWASs provide lessons in observational research, including statistical association and multiple testing. They also provide a way to create hypotheses about biological basis of disease for further study. We will execute a GWAS analysis using data from the Wellcome Trust Case-Control Consortium (WTCCC) on Type 2 Diabetes, Type 1 Diabetes, Bipolar Disorder, Rheumatoid Arthritis, Chron's Disease and Hypertension versus 1,500 controls.\n",
    "\n",
    "Note: There are standard tools and pipelines available to do this but we will do it the old fashioned way!\n",
    "\n",
    "The Data is provided in separate directories under the primary WTCCC directory. The subdirectories are labeled as such:\n",
    "* 58C : control population 1\n",
    "* NBS : control population 2\n",
    "* RA : rheumatoid arthritis\n",
    "* T2D : type 2 diabetes\n",
    "* T1D : type 1 diabetes\n",
    "* CAD : coronary artery disease\n",
    "* HT : hypertension\n",
    "* CD : Crohn’s disease\n",
    "* BD : bipolar disorder\n",
    "\n",
    "All of the GWAS data is broken down per chromosome, 1-22 plus X and is written into the file name, as well as the disease. For example:\n",
    "\n",
    "Affx_gt_ 58C _Chiamo_ 06 .tped.gz\n",
    "\n",
    "Contains GWAS data for the 6th (06) chromosome on control population 1 (58C). The biggest files are chromosome 1 and 2 and smallest is 22.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Each row is a SNP (there are 6207 SNPs measured in Chromosome 22). The first through fourth columns denote the chromosome number, the second is the WTCCC snp_id (not to be confused with the rsid), the start coordinate (0), and the end coordinate. Thereafter, each column is a genotype (where each base is separated by a space character) for that SNP for each individual.\n",
    "\n",
    "Additionally, each subdirectory also contains a ‘snps_info.tar.gz’, which contains the snp_id and rsID for each SNP for each chromosome. These files can enable you to map the WTCCC snp_id to the rsid, which is helpful to find the closest genes to a SNP (for example when you query the GWAS catalog)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/varun/Desktop/BeaverWorks/Week3Public/GWAS\r\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from os import path\n",
    "import csv\n",
    "import numpy as np\n",
    "from time import time\n",
    "!pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Load the test data \n",
    "t1d_22_file = '/Users/varun/Desktop/BeaverWorks/Week3Public/BWSI_100set/T2D/Affx_gt_T2D_Chiamo_22.tped.gz'\n",
    "t1d = pd.read_csv(t1d_22_file, compression='gzip', header=None, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(100, 2003)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Look at the shape of the data frame - you should have 6207 SNPs for 2003 individuals\n",
    "t1d.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22</td>\n",
       "      <td>SNP_A-4234155</td>\n",
       "      <td>0</td>\n",
       "      <td>32056663</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>22</td>\n",
       "      <td>SNP_A-2223191</td>\n",
       "      <td>0</td>\n",
       "      <td>35653934</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>22</td>\n",
       "      <td>SNP_A-4202732</td>\n",
       "      <td>0</td>\n",
       "      <td>32956968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>22</td>\n",
       "      <td>SNP_A-1811354</td>\n",
       "      <td>0</td>\n",
       "      <td>46997362</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>22</td>\n",
       "      <td>SNP_A-1797179</td>\n",
       "      <td>0</td>\n",
       "      <td>32955711</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    0              1  2         3\n",
       "0  22  SNP_A-4234155  0  32056663\n",
       "1  22  SNP_A-2223191  0  35653934\n",
       "2  22  SNP_A-4202732  0  32956968\n",
       "3  22  SNP_A-1811354  0  46997362\n",
       "4  22  SNP_A-1797179  0  32955711"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Look at the top 10 rows\n",
    "# t2d.head(10)\n",
    "t1d_header = t1d.loc[:, :3]\n",
    "t1d_body = t1d.loc[:,4:]\n",
    "t1d_header.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>1498</th>\n",
       "      <th>1499</th>\n",
       "      <th>1500</th>\n",
       "      <th>1501</th>\n",
       "      <th>1502</th>\n",
       "      <th>1503</th>\n",
       "      <th>1504</th>\n",
       "      <th>1505</th>\n",
       "      <th>1506</th>\n",
       "      <th>1507</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22</td>\n",
       "      <td>SNP_A-4234155</td>\n",
       "      <td>0</td>\n",
       "      <td>32056663</td>\n",
       "      <td>C T</td>\n",
       "      <td>C T</td>\n",
       "      <td>C T</td>\n",
       "      <td>T T</td>\n",
       "      <td>C T</td>\n",
       "      <td>T T</td>\n",
       "      <td>...</td>\n",
       "      <td>T T</td>\n",
       "      <td>C C</td>\n",
       "      <td>C T</td>\n",
       "      <td>C T</td>\n",
       "      <td>C T</td>\n",
       "      <td>C C</td>\n",
       "      <td>C T</td>\n",
       "      <td>T T</td>\n",
       "      <td>C T</td>\n",
       "      <td>T T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>22</td>\n",
       "      <td>SNP_A-2223191</td>\n",
       "      <td>0</td>\n",
       "      <td>35653934</td>\n",
       "      <td>T T</td>\n",
       "      <td>C C</td>\n",
       "      <td>C T</td>\n",
       "      <td>T T</td>\n",
       "      <td>C C</td>\n",
       "      <td>C T</td>\n",
       "      <td>...</td>\n",
       "      <td>C T</td>\n",
       "      <td>C T</td>\n",
       "      <td>C T</td>\n",
       "      <td>C T</td>\n",
       "      <td>T T</td>\n",
       "      <td>C T</td>\n",
       "      <td>C T</td>\n",
       "      <td>C C</td>\n",
       "      <td>T T</td>\n",
       "      <td>C C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>22</td>\n",
       "      <td>SNP_A-4202732</td>\n",
       "      <td>0</td>\n",
       "      <td>32956968</td>\n",
       "      <td>G G</td>\n",
       "      <td>A G</td>\n",
       "      <td>A G</td>\n",
       "      <td>G G</td>\n",
       "      <td>A G</td>\n",
       "      <td>A G</td>\n",
       "      <td>...</td>\n",
       "      <td>A G</td>\n",
       "      <td>A G</td>\n",
       "      <td>A G</td>\n",
       "      <td>G G</td>\n",
       "      <td>G G</td>\n",
       "      <td>G G</td>\n",
       "      <td>G G</td>\n",
       "      <td>G G</td>\n",
       "      <td>A G</td>\n",
       "      <td>G G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>22</td>\n",
       "      <td>SNP_A-1811354</td>\n",
       "      <td>0</td>\n",
       "      <td>46997362</td>\n",
       "      <td>A G</td>\n",
       "      <td>A A</td>\n",
       "      <td>A G</td>\n",
       "      <td>A A</td>\n",
       "      <td>A G</td>\n",
       "      <td>G G</td>\n",
       "      <td>...</td>\n",
       "      <td>A A</td>\n",
       "      <td>A G</td>\n",
       "      <td>A A</td>\n",
       "      <td>A A</td>\n",
       "      <td>A G</td>\n",
       "      <td>A G</td>\n",
       "      <td>A A</td>\n",
       "      <td>A A</td>\n",
       "      <td>A G</td>\n",
       "      <td>A A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>22</td>\n",
       "      <td>SNP_A-1797179</td>\n",
       "      <td>0</td>\n",
       "      <td>32955711</td>\n",
       "      <td>G G</td>\n",
       "      <td>A G</td>\n",
       "      <td>A G</td>\n",
       "      <td>G G</td>\n",
       "      <td>A G</td>\n",
       "      <td>A G</td>\n",
       "      <td>...</td>\n",
       "      <td>A G</td>\n",
       "      <td>A G</td>\n",
       "      <td>A G</td>\n",
       "      <td>G G</td>\n",
       "      <td>G G</td>\n",
       "      <td>G G</td>\n",
       "      <td>G G</td>\n",
       "      <td>G G</td>\n",
       "      <td>A G</td>\n",
       "      <td>G G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 1508 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   0              1     2         3    4    5    6    7    8    9     ...  \\\n",
       "0    22  SNP_A-4234155     0  32056663  C T  C T  C T  T T  C T  T T  ...   \n",
       "1    22  SNP_A-2223191     0  35653934  T T  C C  C T  T T  C C  C T  ...   \n",
       "2    22  SNP_A-4202732     0  32956968  G G  A G  A G  G G  A G  A G  ...   \n",
       "3    22  SNP_A-1811354     0  46997362  A G  A A  A G  A A  A G  G G  ...   \n",
       "4    22  SNP_A-1797179     0  32955711  G G  A G  A G  G G  A G  A G  ...   \n",
       "\n",
       "  1498 1499 1500 1501 1502 1503 1504 1505 1506 1507  \n",
       "0  T T  C C  C T  C T  C T  C C  C T  T T  C T  T T  \n",
       "1  C T  C T  C T  C T  T T  C T  C T  C C  T T  C C  \n",
       "2  A G  A G  A G  G G  G G  G G  G G  G G  A G  G G  \n",
       "3  A A  A G  A A  A A  A G  A G  A A  A A  A G  A A  \n",
       "4  A G  A G  A G  G G  G G  G G  G G  G G  A G  G G  \n",
       "\n",
       "[5 rows x 1508 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "control_file = '/Users/varun/Desktop/BeaverWorks/Week3Public/BWSI_100set/58C/Affx_gt_58C_Chiamo_22.tped.gz'\n",
    "control_58c = pd.read_csv(control_file, compression='gzip', header=None, sep='\\t')\n",
    "\n",
    "control_58c_header = control_58c.loc[:, :3]\n",
    "control_58c_body = control_58c.loc[:, 4:]\n",
    "control_58c.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# control_file = '/Users/varun/Desktop/BeaverWorks/Week3Public/BWSI_100set/NBS/Affx_gt_NBS_Chiamo_01.tped.gz'\n",
    "# control_nbs = pd.read_csv(control_file, compression='gzip', header=None, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# control_nbs_header = control_nbs.loc[:, :3]\n",
    "# control_nbs_body = control_nbs.loc[:, 4:]\n",
    "# control_nbs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>SNP</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22</td>\n",
       "      <td>SNP_A-4234155</td>\n",
       "      <td>0</td>\n",
       "      <td>32056663</td>\n",
       "      <td>rs12484156</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>22</td>\n",
       "      <td>SNP_A-2223191</td>\n",
       "      <td>0</td>\n",
       "      <td>35653934</td>\n",
       "      <td>rs909486</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>22</td>\n",
       "      <td>SNP_A-4202732</td>\n",
       "      <td>0</td>\n",
       "      <td>32956968</td>\n",
       "      <td>rs1873230</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>22</td>\n",
       "      <td>SNP_A-1811354</td>\n",
       "      <td>0</td>\n",
       "      <td>46997362</td>\n",
       "      <td>rs7287281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>22</td>\n",
       "      <td>SNP_A-1797179</td>\n",
       "      <td>0</td>\n",
       "      <td>32955711</td>\n",
       "      <td>rs1003935</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    0            SNP  2         3           4\n",
       "0  22  SNP_A-4234155  0  32056663  rs12484156\n",
       "1  22  SNP_A-2223191  0  35653934    rs909486\n",
       "2  22  SNP_A-4202732  0  32956968   rs1873230\n",
       "3  22  SNP_A-1811354  0  46997362   rs7287281\n",
       "4  22  SNP_A-1797179  0  32955711   rs1003935"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp_file = '/Users/varun/Desktop/BeaverWorks/Week3Public/BWSI_100set/58C/snps/snps_22'\n",
    "snp = pd.read_csv(snp_file, header=None, sep='\\t')\n",
    "\n",
    "combined = pd.concat([control_58c_header, snp], axis=1)\n",
    "# combined = combined[:100]\n",
    "\n",
    "control_58c_header = control_58c_header.rename(columns={1:'SNP'})\n",
    "snp = snp.loc[:, 3:]\n",
    "snp = snp.rename(columns={3:'SNP'})\n",
    "combined = control_58c_header.join(snp.set_index('SNP'), on='SNP')\n",
    "combined.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>SNP</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22</td>\n",
       "      <td>SNP_A-4234155</td>\n",
       "      <td>0</td>\n",
       "      <td>32056663</td>\n",
       "      <td>rs12484156</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>22</td>\n",
       "      <td>SNP_A-2223191</td>\n",
       "      <td>0</td>\n",
       "      <td>35653934</td>\n",
       "      <td>rs909486</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>22</td>\n",
       "      <td>SNP_A-4202732</td>\n",
       "      <td>0</td>\n",
       "      <td>32956968</td>\n",
       "      <td>rs1873230</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>22</td>\n",
       "      <td>SNP_A-1811354</td>\n",
       "      <td>0</td>\n",
       "      <td>46997362</td>\n",
       "      <td>rs7287281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>22</td>\n",
       "      <td>SNP_A-1797179</td>\n",
       "      <td>0</td>\n",
       "      <td>32955711</td>\n",
       "      <td>rs1003935</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    0            SNP  2         3           4\n",
       "0  22  SNP_A-4234155  0  32056663  rs12484156\n",
       "1  22  SNP_A-2223191  0  35653934    rs909486\n",
       "2  22  SNP_A-4202732  0  32956968   rs1873230\n",
       "3  22  SNP_A-1811354  0  46997362   rs7287281\n",
       "4  22  SNP_A-1797179  0  32955711   rs1003935"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined = pd.concat([t1d_header, snp], axis=1)\n",
    "# combined = combined[:100]\n",
    "\n",
    "t1d_header = t1d_header.rename(columns={1:'SNP'})\n",
    "combined_t1d = t1d_header.join(snp.set_index('SNP'), on='SNP')\n",
    "combined_t1d.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# row = 0\n",
    "# t1d_1 = t1d.loc[row, 4:]\n",
    "# # print(t1d)\n",
    "\n",
    "# control_58c_1 = control_58c_body.loc[0]\n",
    "# print(\"Snip: \" + str(t1d.loc[row, 1])[6:])\n",
    "# print(\"RSID: \" + str(combined_t1d.loc[row, 4])[2:])\n",
    "# print(\"Type 2 Diabetes:\")\n",
    "# print(computeAlleleFrequency(getAlleleCounts(t1d_1)[0], getAlleleCounts(t1d_1)[1]))\n",
    "# t1_minor = computeAlleleFrequency(getAlleleCounts(t1d_1)[0], getAlleleCounts(t1d_1)[1])\n",
    "# print(1-computeAlleleFrequency(getAlleleCounts(t1d_1)[0], getAlleleCounts(t1d_1)[1]))\n",
    "# t1_major = 1 - t1_minor\n",
    "# print(\"\\nControl:\")\n",
    "# print(computeAlleleFrequency(getAlleleCounts(control_58c_1)[0], getAlleleCounts(control_58c_1)[1]))\n",
    "# control_minor = computeAlleleFrequency(getAlleleCounts(control_58c_1)[0], getAlleleCounts(control_58c_1)[1])\n",
    "# print(1-computeAlleleFrequency(getAlleleCounts(control_58c_1)[0], getAlleleCounts(control_58c_1)[1]))\n",
    "# control_major = 1-control_minor\n",
    "# # print(control)\n",
    "# print(\"\\nOdds Ratio:\")\n",
    "# print(computeOR([t1_minor, t1_major, control_minor, control_major]))\n",
    "\n",
    "# print(\"\\nChi-Squared:\")\n",
    "# print(getPValue([t1_minor, t1_major]))\n",
    "# print(getPValue([control_minor, control_major]))\n",
    "\n",
    "# print(\"\\nHWE:\")\n",
    "# print(HWEChiSq(t1_minor, [t1d_1.to_string().count('C C'), t1d_1.to_string().count('C T'), t1d_1.to_string().count('T T')]))\n",
    "# print(HWEChiSq(control_minor, [control_58c_1.to_string().count('C C'), control_58c_1.to_string().count('C T'), control_58c_1.to_string().count('T T')]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def findAlleles(genotype):\n",
    "    arr = []\n",
    "    shit = []\n",
    "    gen = genotype.str\n",
    "    c = gen.count('C').sum()\n",
    "    t = gen.count('T').sum()\n",
    "    a = gen.count('A').sum()\n",
    "    g = gen.count('G').sum()\n",
    "    \n",
    "    if c>0:\n",
    "        arr.append('C')\n",
    "        shit.append(c)\n",
    "    if t>0:\n",
    "        arr.append('T')\n",
    "        shit.append(t)\n",
    "    if a>0:\n",
    "        arr.append('A')\n",
    "        shit.append(a)\n",
    "    if g>0:\n",
    "        arr.append('G')\n",
    "        shit.append(g)\n",
    "        \n",
    "    if len(arr) == 1:\n",
    "        return None\n",
    "    elif shit[0] > shit[1]:\n",
    "        return [arr[1], arr[0]]\n",
    "    else:\n",
    "        return arr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "## function to load the data file\n",
    "def loadFile (disease, chrom_num):\n",
    "    ## construct your file name based on the disease and chrom_number\n",
    "    if(chrom_num<10):\n",
    "        fileName = '/Users/varun/Desktop/BeaverWorks/Week3Public/BWSI_100set/' + str(disease) + '/Affx_gt_' + str(disease) + '_Chiamo_0' + str(chrom_num) + '.tped.gz'   \n",
    "    else:\n",
    "        fileName = '/Users/varun/Desktop/BeaverWorks/Week3Public/BWSI_100set/' + str(disease) + '/Affx_gt_' + str(disease) + '_Chiamo_' + str(chrom_num) + '.tped.gz'\n",
    "    dat = pd.read_csv(fileName, compression='gzip', header=None, sep='\\t')\n",
    "    return (dat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "## compute Allele Frequency for the minor allele\n",
    "def computeAlleleFrequency (f_C, f_T):\n",
    "    \n",
    "    \"\"\"\n",
    "    f_C = minor allele count\n",
    "    f_T = major allele count\n",
    "    minor_allele_frequency = f_C/ (f_C+f_T) \n",
    "    \"\"\"\n",
    "    minor_allele_frequency = f_C/ (f_C+f_T) \n",
    "    return (minor_allele_frequency)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getAlleleCounts (genotype, arr):\n",
    "    \"\"\"\n",
    "    genotype = \"C C \tC T \tC T \tC T \tT T\"\n",
    "    Allele frequency:\n",
    "    count the \"Cs\", and the \"Ts\"\n",
    "    f_C = # of Cs\n",
    "    f_T = # of T's\n",
    "    \"\"\"\n",
    "\n",
    "    \n",
    "    return np.array([genotype.str.count(arr[0]).sum(), genotype.str.count(arr[1]).sum()])\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Hardy Weinberg Equilibrium"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Function takes as input the minor allele frequency (p) and the population genotype counts\n",
    "## returns the p-value from the chisq test\n",
    "import scipy\n",
    "def HWEChiSq (p, populationGenotypeCounts):\n",
    "    \"\"\"\n",
    "        |  CC | CT  | TT |\n",
    "    BP  | 270 | 957 | 771|\n",
    "    compute the HWE\n",
    "      -- p = frequency of the C allele\n",
    "      -- q = 1-p = frequency of the T allele\n",
    "      -- if the population genotype is in Hardy Weinberg Equilibrium, we would expect the genotype frequencies to be\n",
    "          CC = p^2*N \n",
    "          CT = 2pq*N \n",
    "          TT = q^2*N \n",
    "      -- to compute the deviation from HWE\n",
    "         (observed - expected)^2 / expected\n",
    "      -- do a chi.squared test to check if the deviation is significant\n",
    "    \"\"\"\n",
    "    total = sum(populationGenotypeCounts)\n",
    "\n",
    "    q = 1-p\n",
    "    \n",
    "#     observed = populationGenotypeCounts[0]\n",
    "    CC_expected = p**2*total\n",
    "    \n",
    "#     CC = (observed-expected)**2/expected\n",
    "    \n",
    "    \n",
    "#     observed = populationGenotypeCounts[1]\n",
    "    CT_expected = 2*p*q*total\n",
    "    \n",
    "#     CT = (observed-expected)**2/expected\n",
    "    \n",
    "#     observed = populationGenotypeCounts[2]\n",
    "    TT_expected = q**2*total\n",
    "    \n",
    "#     TT = (observed-expected)**2/expected\n",
    "    \n",
    "   \n",
    "#     from scipy.stats.distributions import chi2\n",
    "#     return chi2.sf(CC+CT+TT, 1)\n",
    "    test_result = scipy.stats.chisquare(f_obs=populationGenotypeCounts, f_exp=np.array([CC_expected, CT_expected, TT_expected]), ddof=1, axis=0)\n",
    "    return test_result[1]\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Compute the Odds Ratio\n",
    "## takes as input a 2X2  confusion matrix\n",
    "## returns the odds ratio\n",
    "def computeOR (confusionMatrix):\n",
    "    ad = confusionMatrix[0][0] * confusionMatrix[1][1]\n",
    "    bc = confusionMatrix[1][0] * confusionMatrix[0][1]\n",
    "    return ad/bc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Execute a chisq test to determine if the odds ratio is significant\n",
    "## return the p-value from the chisq test\n",
    "from scipy.stats import chisquare\n",
    "def getPValue (confusionMatrix):\n",
    "    return chisquare(confusionMatrix).pvalue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "## plot the data\n",
    "## manhattan plot\n",
    "## the X axis represents the chromosomes \n",
    "## the y-axis is the -log(p.value)\n",
    "def manhattanPlot (diseaseResults):\n",
    "    \"\"\"\n",
    "    group data by chromosomes, SNP before plotting\n",
    "    \"\"\"\n",
    "    return "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Write code to run the GWAS\n",
    "\n",
    "## Write a function that will take the following inputs:\n",
    "## (a) control files - 58C and NBS\n",
    "## (b) the disease - T2D, T1D, HT, BD, CD, RA\n",
    "## (c) the chromosome number\n",
    "\n",
    "## Your function should output a csv file with the following information\n",
    "## SNP Id, RSID, CHROMOSOME NUMBER, MINOR ALLELE, MAJOR ALLELE, MINOR ALLELE FREQUENCY, MAJOR ALLELE FREQUENCY, \n",
    "## ODDS RATIO (MINOR vs MAJOR ALLELE), P-Value for ODDS RATIO, Chi square for the ODDS RATIO, HWE deviation P-VALUE\n",
    "## \n",
    "from scipy.stats import chi2_contingency\n",
    "\n",
    "def gwas (disease, chrom_num, row, t1d, control_58c_body, combined_t1d):\n",
    "    \"\"\"\n",
    "    1. Load the control file for the specified chromosome number. Since there are two control populations, \n",
    "       it would be best to combine the two controls together\n",
    "    2. Load the disease file for the given chromosome number\n",
    "    3. Split the header from the SNP data for the controls and the disease\n",
    "    3. Load the SNP file -- this gives you the mapping between the WTCCC SNP ids and the RSID\n",
    "    4. For each SNP in the disease file and its matching SNP in the control file:\n",
    "       -- to make it easier, before you begin do a check for zygosity. If either the controls or the disease \n",
    "       are homozygous, skip the SNP\n",
    "       genotype CC, TT : skip\n",
    "       genotype CC CT TT : run the following analysis\n",
    "       a. get the minor and major allele counts and compute the minor allele frequency for the disease population\n",
    "       b. get the minor and major allele counts and compute the minor allele frequency for the control population\n",
    "       c. conduct the allelic test -- compute the odds ratios to test the frequency of the minor allele \n",
    "          in the disease population compared to the control population\n",
    "          -- you can build a confusion matrix based on the counts that you have calculated in a and b\n",
    "          -- compute your odds ratios based on the minor allele of the control population\n",
    "          \n",
    "          |                  | C    | T    | \n",
    "          |------------------+ ---  + ---  +\n",
    "          | Bipolar Disorder | 1529 | 2469 |\n",
    "          |------------------+----  + ---  +\n",
    "          | Healthy Controls | 2270 | 3738 |\n",
    "          |------------------+----- +----- +\n",
    "          \n",
    "          (Test : your OR for above should be :1.019)\n",
    "       d. Do a chi square test to test the significant and record the p-value\n",
    "       d. for each SNP calculate the deviation from Hardy-Weinberg equilibrium\n",
    "       e. create an output row containing:\n",
    "       SNP Id, RSID, CHROMOSOME NUMBER, MINOR ALLELE, MAJOR ALLELE, MINOR ALLELE FREQUENCY, \n",
    "       MAJOR ALLELE FREQUENCY, ODDS RATIO (MINOR vs MAJOR ALLELE), \n",
    "       P-Value for ODDS RATIO, CHI_SQ for the ODDS RATIO, HWE deviation P-VALUE  \n",
    "    5. Save the result file for the chromosome\n",
    "    \"\"\"\n",
    "    \n",
    "#     dataName = \"LittleBoy\"\n",
    "#     path = \"\"\n",
    "#     if(dataName=='Japan'):\n",
    "#         path = \"/Users/varun/Desktop/BeaverWorks/Week3Public/BWSI_100set/\"\n",
    "#     elif(dataName=='LittleBoy'):\n",
    "#         path = \"/Users/varun/Desktop/BeaverWorks/Week3Public/BWSI_100set/LittleBoy\"\n",
    "#     elif(dataName=='FatMan'):\n",
    "#         path = \"/Users/varun/Desktop/BeaverWorks/Week3Public/BWSI_100set/FatMan\"\n",
    "        \n",
    "    \n",
    "#     if(chrom_num<10):\n",
    "#         fileName = path + '/' + str(disease) + '/Affx_gt_' + str(disease) + '_Chiamo_0' + str(chrom_num) + '.tped.gz'   \n",
    "#     else:\n",
    "#         fileName = path + '/' + str(disease) + '/Affx_gt_' + str(disease) + '_Chiamo_' + str(chrom_num) + '.tped.gz'\n",
    "#     data = pd.read_csv(fileName, compression='gzip', header=None, sep='\\t')\n",
    "\n",
    "#     if chrom_num < 10:\n",
    "#         chrom_num = \"0\"+str(chrom_num)\n",
    "#     control_file = path + '/58C/Affx_gt_58C_Chiamo_' + str(chrom_num) + '.tped.gz'\n",
    "#     control_58c = pd.read_csv(control_file, compression='gzip', header=None, sep='\\t')\n",
    "\n",
    "\n",
    "#     control_58c_header = control_58c.loc[:, :3]\n",
    "#     control_58c_body = control_58c.loc[:, 4:]\n",
    "\n",
    "#     ## Load the test data \n",
    "#     #     t1d_22_file = '/Users/varun/Desktop/BeaverWorks/Week3Public/BWSI_100set/T1D/Affx_gt_T2D_Chiamo_22.tped.gz'\n",
    "#     t1d = data\n",
    "\n",
    "#     t1d_header = data.loc[:, :3]\n",
    "#     t1d_body = data.loc[:,4:]\n",
    "\n",
    "#     snp_file = path + '/58C/snps/snps_'+str(chrom_num)\n",
    "#     snp = pd.read_csv(snp_file, header=None, sep='\\t')\n",
    "\n",
    "\n",
    "#     combined = pd.concat([control_58c_header, snp], axis=1)\n",
    "    t1d_1 = t1d.loc[row, 4:]\n",
    "    control_58c_1 = control_58c_body.loc[row]\n",
    "    control_58c_string = control_58c_1.to_string()\n",
    "    t1d_string = t1d_1.to_string()\n",
    "    if(findAlleles(t1d_1)!=None and findAlleles(control_58c_1)!=None):\n",
    "        snp = int(str(t1d.loc[row, 1])[6:])\n",
    "        rsid=int(str(combined_t1d.loc[row, 4])[2:])\n",
    "        minorAllele = findAlleles(t1d_1)[0]\n",
    "        majorAllele = findAlleles(t1d_1)[1]\n",
    "        t1_minor = computeAlleleFrequency(getAlleleCounts(t1d_1, findAlleles(t1d_1))[0], getAlleleCounts(t1d_1, findAlleles(t1d_1))[1])\n",
    "        if(t1_minor < 0.01):\n",
    "            return pd.Series([])\n",
    "        t1_major = 1 - t1_minor\n",
    "        control_minor = computeAlleleFrequency(getAlleleCounts(control_58c_1, findAlleles(control_58c_1))[0], getAlleleCounts(control_58c_1, findAlleles(control_58c_1))[1])\n",
    "        if(control_minor < 0.01):\n",
    "            return pd.Series([])\n",
    "        HWEControl=HWEChiSq(control_minor, [control_58c_string.count(minorAllele + ' ' + minorAllele), control_58c_string.count(minorAllele + ' ' + majorAllele), control_58c_string.count(majorAllele + ' ' + majorAllele)])\n",
    "        if(HWEControl < 0.05):\n",
    "            return pd.Series([])\n",
    "        control_major = 1-control_minor\n",
    "        oddsRatio=computeOR([[t1_minor, t1_major], [control_minor, control_major]])\n",
    "        oddsRatioP=chi2_contingency(np.array([[t1d_string.count(minorAllele),t1d_string.count(majorAllele)], [control_58c_string.count(minorAllele),control_58c_string.count(majorAllele)]]))[1]\n",
    "        HWEDisease=HWEChiSq(t1_minor, [t1d_string.count(minorAllele + ' ' + minorAllele), t1d_string.count(minorAllele + ' ' + majorAllele), t1d_string.count(majorAllele + ' ' + majorAllele)])\n",
    "        logOddsRatioP = np.log10(oddsRatioP)*-1\n",
    "        logHWEControl= np.log10(HWEControl)*-1\n",
    "        return pd.Series([chrom_num, snp, rsid, minorAllele, majorAllele, control_minor, control_major, t1_minor, t1_major, oddsRatio, oddsRatioP, HWEControl, logOddsRatioP, logHWEControl])\n",
    "    else:\n",
    "        return pd.Series([])\n",
    "# arr = gwas('T1D', 22)\n",
    "# # print(len(arr[0])\n",
    "# df = pd.DataFrame({\"Chromosome\":arr[0],\n",
    "#                   \"SNP\": arr[1],\n",
    "#                    \"Minor Allele\": arr[9],\n",
    "#                    \"Major Allele\": arr[10],\n",
    "#                   \"Minor Allele Control Frequency\": arr[2],\n",
    "#                   \"Major Allele Control Frequency\": arr[3],\n",
    "#                   \"Minor Allele Disease Frequency\": arr[4],\n",
    "#                   \"Major Allele Disease Frequency\": arr[5],\n",
    "#                   \"Odds Ratio\": arr[6],\n",
    "#                   \"Odds Ratio p-value\": arr[7],\n",
    "#                   \"HWE\": arr[8]})\n",
    "# df.to_csv(\"data.csv\", sep='\\t')\n",
    "# print(\"Done!\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "# print(len(arr[1]))\n",
    "# print(len(arr[8]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/javascript": [
       "IPython.OutputArea.prototype._should_scroll = function(lines) {\n",
       "    return false;\n",
       "}"
      ],
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%javascript\n",
    "IPython.OutputArea.prototype._should_scroll = function(lines) {\n",
    "    return false;\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chromosome 1\n",
      "SNP 0\n",
      "SNP 1\n",
      "SNP 2\n",
      "SNP 3\n",
      "SNP 4\n",
      "SNP 5\n",
      "SNP 6\n",
      "SNP 7\n",
      "SNP 8\n",
      "SNP 9\n",
      "SNP 10\n",
      "SNP 11\n",
      "SNP 12\n",
      "SNP 13\n",
      "SNP 14\n",
      "SNP 15\n",
      "SNP 16\n",
      "SNP 17\n",
      "SNP 18\n",
      "SNP 19\n",
      "SNP 20\n",
      "SNP 21\n",
      "SNP 22\n",
      "SNP 23\n",
      "SNP 24\n",
      "SNP 25\n",
      "SNP 26\n",
      "SNP 27\n",
      "SNP 28\n",
      "SNP 29\n",
      "SNP 30\n",
      "SNP 31\n",
      "SNP 32\n",
      "SNP 33\n",
      "SNP 34\n",
      "SNP 35\n",
      "SNP 36\n",
      "SNP 37\n",
      "SNP 38\n",
      "SNP 39\n",
      "SNP 40\n",
      "SNP 41\n",
      "SNP 42\n",
      "SNP 43\n",
      "SNP 44\n",
      "SNP 45\n",
      "SNP 46\n",
      "SNP 47\n",
      "SNP 48\n",
      "SNP 49\n",
      "SNP 50\n",
      "SNP 51\n",
      "SNP 52\n",
      "SNP 53\n",
      "SNP 54\n",
      "SNP 55\n",
      "SNP 56\n",
      "SNP 57\n",
      "SNP 58\n",
      "SNP 59\n",
      "SNP 60\n",
      "SNP 61\n",
      "SNP 62\n",
      "SNP 63\n",
      "SNP 64\n",
      "SNP 65\n",
      "SNP 66\n",
      "SNP 67\n",
      "SNP 68\n",
      "SNP 69\n",
      "SNP 70\n",
      "SNP 71\n",
      "SNP 72\n",
      "SNP 73\n",
      "SNP 74\n",
      "SNP 75\n",
      "SNP 76\n",
      "SNP 77\n",
      "SNP 78\n",
      "SNP 79\n",
      "SNP 80\n",
      "SNP 81\n",
      "SNP 82\n",
      "SNP 83\n",
      "SNP 84\n",
      "SNP 85\n",
      "SNP 86\n",
      "SNP 87\n",
      "SNP 88\n",
      "SNP 89\n",
      "SNP 90\n",
      "SNP 91\n",
      "SNP 92\n",
      "SNP 93\n",
      "SNP 94\n",
      "SNP 95\n",
      "SNP 96\n",
      "SNP 97\n",
      "SNP 98\n",
      "SNP 99\n",
      "SNP 100\n",
      "SNP 101\n",
      "SNP 102\n",
      "SNP 103\n",
      "SNP 104\n",
      "SNP 105\n",
      "SNP 106\n",
      "SNP 107\n",
      "SNP 108\n",
      "SNP 109\n",
      "SNP 110\n",
      "SNP 111\n",
      "SNP 112\n",
      "SNP 113\n",
      "SNP 114\n",
      "SNP 115\n",
      "SNP 116\n",
      "SNP 117\n",
      "SNP 118\n",
      "SNP 119\n",
      "SNP 120\n",
      "SNP 121\n",
      "SNP 122\n",
      "SNP 123\n",
      "SNP 124\n",
      "SNP 125\n",
      "SNP 126\n",
      "SNP 127\n",
      "SNP 128\n",
      "SNP 129\n",
      "SNP 130\n",
      "SNP 131\n",
      "SNP 132\n",
      "SNP 133\n",
      "SNP 134\n",
      "SNP 135\n",
      "SNP 136\n",
      "SNP 137\n",
      "SNP 138\n",
      "SNP 139\n",
      "SNP 140\n",
      "SNP 141\n",
      "SNP 142\n",
      "SNP 143\n",
      "SNP 144\n",
      "SNP 145\n",
      "SNP 146\n",
      "SNP 147\n",
      "SNP 148\n",
      "SNP 149\n",
      "SNP 150\n",
      "SNP 151\n",
      "SNP 152\n",
      "SNP 153\n",
      "SNP 154\n",
      "SNP 155\n",
      "SNP 156\n",
      "SNP 157\n",
      "SNP 158\n",
      "SNP 159\n",
      "SNP 160\n",
      "SNP 161\n",
      "SNP 162\n",
      "SNP 163\n",
      "SNP 164\n",
      "SNP 165\n",
      "SNP 166\n",
      "SNP 167\n",
      "SNP 168\n",
      "SNP 169\n",
      "SNP 170\n",
      "SNP 171\n",
      "SNP 172\n",
      "SNP 173\n",
      "SNP 174\n",
      "SNP 175\n",
      "SNP 176\n",
      "SNP 177\n",
      "SNP 178\n",
      "SNP 179\n",
      "SNP 180\n",
      "SNP 181\n",
      "SNP 182\n",
      "SNP 183\n",
      "SNP 184\n",
      "SNP 185\n",
      "SNP 186\n",
      "SNP 187\n",
      "SNP 188\n",
      "SNP 189\n",
      "SNP 190\n",
      "SNP 191\n",
      "SNP 192\n",
      "SNP 193\n",
      "SNP 194\n",
      "SNP 195\n",
      "SNP 196\n",
      "SNP 197\n",
      "SNP 198\n",
      "SNP 199\n",
      "SNP 200\n",
      "SNP 201\n",
      "SNP 202\n",
      "SNP 203\n",
      "SNP 204\n",
      "SNP 205\n",
      "SNP 206\n",
      "SNP 207\n",
      "SNP 208\n",
      "SNP 209\n",
      "SNP 210\n",
      "SNP 211\n",
      "SNP 212\n",
      "SNP 213\n",
      "SNP 214\n",
      "SNP 215\n",
      "SNP 216\n",
      "SNP 217\n",
      "SNP 218\n",
      "SNP 219\n",
      "SNP 220\n",
      "SNP 221\n",
      "SNP 222\n",
      "SNP 223\n",
      "SNP 224\n",
      "SNP 225\n",
      "SNP 226\n",
      "SNP 227\n",
      "SNP 228\n",
      "SNP 229\n",
      "SNP 230\n",
      "SNP 231\n",
      "SNP 232\n",
      "SNP 233\n",
      "SNP 234\n",
      "SNP 235\n",
      "SNP 236\n",
      "SNP 237\n",
      "SNP 238\n",
      "SNP 239\n",
      "SNP 240\n",
      "SNP 241\n",
      "SNP 242\n",
      "SNP 243\n",
      "SNP 244\n",
      "SNP 245\n",
      "SNP 246\n",
      "SNP 247\n",
      "SNP 248\n",
      "SNP 249\n",
      "SNP 250\n",
      "SNP 251\n",
      "SNP 252\n",
      "SNP 253\n",
      "SNP 254\n",
      "SNP 255\n",
      "SNP 256\n",
      "SNP 257\n",
      "SNP 258\n",
      "SNP 259\n",
      "SNP 260\n",
      "SNP 261\n",
      "SNP 262\n",
      "SNP 263\n",
      "SNP 264\n",
      "SNP 265\n",
      "SNP 266\n",
      "SNP 267\n",
      "SNP 268\n",
      "SNP 269\n",
      "SNP 270\n",
      "SNP 271\n",
      "SNP 272\n",
      "SNP 273\n",
      "SNP 274\n",
      "SNP 275\n",
      "SNP 276\n",
      "SNP 277\n",
      "SNP 278\n",
      "SNP 279\n",
      "SNP 280\n",
      "SNP 281\n",
      "SNP 282\n",
      "SNP 283\n",
      "SNP 284\n",
      "SNP 285\n",
      "SNP 286\n",
      "SNP 287\n",
      "SNP 288\n",
      "SNP 289\n",
      "SNP 290\n",
      "SNP 291\n",
      "SNP 292\n",
      "SNP 293\n",
      "SNP 294\n",
      "SNP 295\n",
      "SNP 296\n",
      "SNP 297\n",
      "SNP 298\n",
      "SNP 299\n",
      "SNP 300\n",
      "SNP 301\n",
      "SNP 302\n",
      "SNP 303\n",
      "SNP 304\n",
      "SNP 305\n",
      "SNP 306\n",
      "SNP 307\n",
      "SNP 308\n",
      "SNP 309\n",
      "SNP 310\n",
      "SNP 311\n",
      "SNP 312\n",
      "SNP 313\n",
      "SNP 314\n",
      "SNP 315\n",
      "SNP 316\n",
      "SNP 317\n",
      "SNP 318\n",
      "SNP 319\n",
      "SNP 320\n",
      "SNP 321\n",
      "SNP 322\n",
      "SNP 323\n",
      "SNP 324\n",
      "SNP 325\n",
      "SNP 326\n",
      "SNP 327\n",
      "SNP 328\n",
      "SNP 329\n",
      "SNP 330\n",
      "SNP 331\n",
      "SNP 332\n",
      "SNP 333\n",
      "SNP 334\n",
      "SNP 335\n",
      "SNP 336\n",
      "SNP 337\n",
      "SNP 338\n",
      "SNP 339\n",
      "SNP 340\n",
      "SNP 341\n",
      "SNP 342\n",
      "SNP 343\n",
      "SNP 344\n",
      "SNP 345\n",
      "SNP 346\n",
      "SNP 347\n",
      "SNP 348\n",
      "SNP 349\n",
      "SNP 350\n",
      "SNP 351\n",
      "SNP 352\n",
      "SNP 353\n",
      "SNP 354\n",
      "SNP 355\n",
      "SNP 356\n",
      "SNP 357\n",
      "SNP 358\n",
      "SNP 359\n",
      "SNP 360\n",
      "SNP 361\n",
      "SNP 362\n",
      "SNP 363\n",
      "SNP 364\n",
      "SNP 365\n",
      "SNP 366\n",
      "SNP 367\n",
      "SNP 368\n",
      "SNP 369\n",
      "SNP 370\n",
      "SNP 371\n",
      "SNP 372\n",
      "SNP 373\n",
      "SNP 374\n",
      "SNP 375\n",
      "SNP 376\n",
      "SNP 377\n",
      "SNP 378\n",
      "SNP 379\n",
      "SNP 380\n",
      "SNP 381\n",
      "SNP 382\n",
      "SNP 383\n",
      "SNP 384\n",
      "SNP 385\n",
      "SNP 386\n",
      "SNP 387\n",
      "SNP 388\n",
      "SNP 389\n",
      "SNP 390\n",
      "SNP 391\n",
      "SNP 392\n",
      "SNP 393\n",
      "SNP 394\n",
      "SNP 395\n",
      "SNP 396\n",
      "SNP 397\n",
      "SNP 398\n",
      "SNP 399\n",
      "SNP 400\n",
      "SNP 401\n",
      "SNP 402\n",
      "SNP 403\n",
      "SNP 404\n",
      "SNP 405\n",
      "SNP 406\n",
      "SNP 407\n",
      "SNP 408\n",
      "SNP 409\n",
      "SNP 410\n",
      "SNP 411\n",
      "SNP 412\n",
      "SNP 413\n",
      "SNP 414\n",
      "SNP 415\n",
      "SNP 416\n",
      "SNP 417\n",
      "SNP 418\n",
      "SNP 419\n",
      "SNP 420\n",
      "SNP 421\n",
      "SNP 422\n",
      "SNP 423\n",
      "SNP 424\n",
      "SNP 425\n",
      "SNP 426\n",
      "SNP 427\n",
      "SNP 428\n",
      "SNP 429\n",
      "SNP 430\n",
      "SNP 431\n",
      "SNP 432\n",
      "SNP 433\n",
      "SNP 434\n",
      "SNP 435\n",
      "SNP 436\n",
      "SNP 437\n",
      "SNP 438\n",
      "SNP 439\n",
      "SNP 440\n",
      "SNP 441\n",
      "SNP 442\n",
      "SNP 443\n",
      "SNP 444\n",
      "SNP 445\n",
      "SNP 446\n",
      "SNP 447\n",
      "SNP 448\n",
      "SNP 449\n",
      "SNP 450\n",
      "SNP 451\n",
      "SNP 452\n",
      "SNP 453\n",
      "SNP 454\n",
      "SNP 455\n",
      "SNP 456\n",
      "SNP 457\n",
      "SNP 458\n",
      "SNP 459\n",
      "SNP 460\n",
      "SNP 461\n",
      "SNP 462\n",
      "SNP 463\n",
      "SNP 464\n",
      "SNP 465\n",
      "SNP 466\n",
      "SNP 467\n",
      "SNP 468\n",
      "SNP 469\n",
      "SNP 470\n",
      "SNP 471\n",
      "SNP 472\n",
      "SNP 473\n",
      "SNP 474\n",
      "SNP 475\n",
      "SNP 476\n",
      "SNP 477\n",
      "SNP 478\n",
      "SNP 479\n",
      "SNP 480\n",
      "SNP 481\n",
      "SNP 482\n",
      "SNP 483\n",
      "SNP 484\n",
      "SNP 485\n",
      "SNP 486\n",
      "SNP 487\n",
      "SNP 488\n",
      "SNP 489\n",
      "SNP 490\n",
      "SNP 491\n",
      "SNP 492\n",
      "SNP 493\n",
      "SNP 494\n",
      "SNP 495\n",
      "SNP 496\n",
      "SNP 497\n",
      "SNP 498\n",
      "SNP 499\n",
      "SNP 500\n",
      "SNP 501\n",
      "SNP 502\n",
      "SNP 503\n",
      "SNP 504\n",
      "SNP 505\n",
      "SNP 506\n",
      "SNP 507\n",
      "SNP 508\n",
      "SNP 509\n",
      "SNP 510\n",
      "SNP 511\n",
      "SNP 512\n",
      "SNP 513\n",
      "SNP 514\n",
      "SNP 515\n",
      "SNP 516\n",
      "SNP 517\n",
      "SNP 518\n",
      "SNP 519\n",
      "SNP 520\n",
      "SNP 521\n",
      "SNP 522\n",
      "SNP 523\n",
      "SNP 524\n",
      "SNP 525\n",
      "SNP 526\n",
      "SNP 527\n",
      "SNP 528\n",
      "SNP 529\n",
      "SNP 530\n",
      "SNP 531\n",
      "SNP 532\n",
      "SNP 533\n",
      "SNP 534\n",
      "SNP 535\n",
      "SNP 536\n",
      "SNP 537\n",
      "SNP 538\n",
      "SNP 539\n",
      "SNP 540\n",
      "SNP 541\n",
      "SNP 542\n",
      "SNP 543\n",
      "SNP 544\n",
      "SNP 545\n",
      "SNP 546\n",
      "SNP 547\n",
      "SNP 548\n",
      "SNP 549\n",
      "SNP 550\n",
      "SNP 551\n",
      "SNP 552\n",
      "SNP 553\n",
      "SNP 554\n",
      "SNP 555\n",
      "SNP 556\n",
      "SNP 557\n",
      "SNP 558\n",
      "SNP 559\n",
      "SNP 560\n",
      "SNP 561\n",
      "SNP 562\n",
      "SNP 563\n",
      "SNP 564\n",
      "SNP 565\n",
      "SNP 566\n",
      "SNP 567\n",
      "SNP 568\n",
      "SNP 569\n",
      "SNP 570\n",
      "SNP 571\n",
      "SNP 572\n",
      "SNP 573\n",
      "SNP 574\n",
      "SNP 575\n",
      "SNP 576\n",
      "SNP 577\n",
      "SNP 578\n",
      "SNP 579\n",
      "SNP 580\n",
      "SNP 581\n",
      "SNP 582\n",
      "SNP 583\n",
      "SNP 584\n",
      "SNP 585\n",
      "SNP 586\n",
      "SNP 587\n",
      "SNP 588\n",
      "SNP 589\n",
      "SNP 590\n",
      "SNP 591\n",
      "SNP 592\n",
      "SNP 593\n",
      "SNP 594\n",
      "SNP 595\n",
      "SNP 596\n",
      "SNP 597\n",
      "SNP 598\n",
      "SNP 599\n",
      "SNP 600\n",
      "SNP 601\n",
      "SNP 602\n",
      "SNP 603\n",
      "SNP 604\n",
      "SNP 605\n",
      "SNP 606\n",
      "SNP 607\n",
      "SNP 608\n",
      "SNP 609\n",
      "SNP 610\n",
      "SNP 611\n",
      "SNP 612\n",
      "SNP 613\n",
      "SNP 614\n",
      "SNP 615\n",
      "SNP 616\n",
      "SNP 617\n",
      "SNP 618\n",
      "SNP 619\n",
      "SNP 620\n",
      "SNP 621\n",
      "SNP 622\n",
      "SNP 623\n",
      "SNP 624\n",
      "SNP 625\n",
      "SNP 626\n",
      "SNP 627\n",
      "SNP 628\n",
      "SNP 629\n",
      "SNP 630\n",
      "SNP 631\n",
      "SNP 632\n",
      "SNP 633\n",
      "SNP 634\n",
      "SNP 635\n",
      "SNP 636\n",
      "SNP 637\n",
      "SNP 638\n",
      "SNP 639\n",
      "SNP 640\n",
      "SNP 641\n",
      "SNP 642\n",
      "SNP 643\n",
      "SNP 644\n",
      "SNP 645\n",
      "SNP 646\n",
      "SNP 647\n",
      "SNP 648\n",
      "SNP 649\n",
      "SNP 650\n",
      "SNP 651\n",
      "SNP 652\n",
      "SNP 653\n",
      "SNP 654\n",
      "SNP 655\n",
      "SNP 656\n",
      "SNP 657\n",
      "SNP 658\n",
      "SNP 659\n",
      "SNP 660\n",
      "SNP 661\n",
      "SNP 662\n",
      "SNP 663\n",
      "SNP 664\n",
      "SNP 665\n",
      "SNP 666\n",
      "SNP 667\n",
      "SNP 668\n",
      "SNP 669\n",
      "SNP 670\n",
      "SNP 671\n",
      "SNP 672\n",
      "SNP 673\n",
      "SNP 674\n",
      "SNP 675\n",
      "SNP 676\n",
      "SNP 677\n",
      "SNP 678\n",
      "SNP 679\n",
      "SNP 680\n",
      "SNP 681\n",
      "SNP 682\n",
      "SNP 683\n",
      "SNP 684\n",
      "SNP 685\n",
      "SNP 686\n",
      "SNP 687\n",
      "SNP 688\n",
      "SNP 689\n",
      "SNP 690\n",
      "SNP 691\n",
      "SNP 692\n",
      "SNP 693\n",
      "SNP 694\n",
      "SNP 695\n",
      "SNP 696\n",
      "SNP 697\n",
      "SNP 698\n",
      "SNP 699\n",
      "SNP 700\n",
      "SNP 701\n",
      "SNP 702\n",
      "SNP 703\n",
      "SNP 704\n",
      "SNP 705\n",
      "SNP 706\n",
      "SNP 707\n",
      "SNP 708\n",
      "SNP 709\n",
      "SNP 710\n",
      "SNP 711\n",
      "SNP 712\n",
      "SNP 713\n",
      "SNP 714\n",
      "SNP 715\n",
      "SNP 716\n",
      "SNP 717\n",
      "SNP 718\n",
      "SNP 719\n",
      "SNP 720\n",
      "SNP 721\n",
      "SNP 722\n",
      "SNP 723\n",
      "SNP 724\n",
      "SNP 725\n",
      "SNP 726\n",
      "SNP 727\n",
      "SNP 728\n",
      "SNP 729\n",
      "SNP 730\n",
      "SNP 731\n",
      "SNP 732\n",
      "SNP 733\n",
      "SNP 734\n",
      "SNP 735\n",
      "SNP 736\n",
      "SNP 737\n",
      "SNP 738\n",
      "SNP 739\n",
      "SNP 740\n",
      "SNP 741\n",
      "SNP 742\n",
      "SNP 743\n",
      "SNP 744\n",
      "SNP 745\n",
      "SNP 746\n",
      "SNP 747\n",
      "SNP 748\n",
      "SNP 749\n",
      "SNP 750\n",
      "SNP 751\n",
      "SNP 752\n",
      "SNP 753\n",
      "SNP 754\n",
      "SNP 755\n",
      "SNP 756\n",
      "SNP 757\n",
      "SNP 758\n",
      "SNP 759\n",
      "SNP 760\n",
      "SNP 761\n",
      "SNP 762\n",
      "SNP 763\n",
      "SNP 764\n",
      "SNP 765\n",
      "SNP 766\n",
      "SNP 767\n",
      "SNP 768\n",
      "SNP 769\n",
      "SNP 770\n",
      "SNP 771\n",
      "SNP 772\n",
      "SNP 773\n",
      "SNP 774\n",
      "SNP 775\n",
      "SNP 776\n",
      "SNP 777\n",
      "SNP 778\n",
      "SNP 779\n",
      "SNP 780\n",
      "SNP 781\n",
      "SNP 782\n",
      "SNP 783\n",
      "SNP 784\n",
      "SNP 785\n",
      "SNP 786\n",
      "SNP 787\n",
      "SNP 788\n",
      "SNP 789\n",
      "SNP 790\n",
      "SNP 791\n",
      "SNP 792\n",
      "SNP 793\n",
      "SNP 794\n",
      "SNP 795\n",
      "SNP 796\n",
      "SNP 797\n",
      "SNP 798\n",
      "SNP 799\n",
      "SNP 800\n",
      "SNP 801\n",
      "SNP 802\n",
      "SNP 803\n",
      "SNP 804\n",
      "SNP 805\n",
      "SNP 806\n",
      "SNP 807\n",
      "SNP 808\n",
      "SNP 809\n",
      "SNP 810\n",
      "SNP 811\n",
      "SNP 812\n",
      "SNP 813\n",
      "SNP 814\n",
      "SNP 815\n",
      "SNP 816\n",
      "SNP 817\n",
      "SNP 818\n",
      "SNP 819\n",
      "SNP 820\n",
      "SNP 821\n",
      "SNP 822\n",
      "SNP 823\n",
      "SNP 824\n",
      "SNP 825\n",
      "SNP 826\n",
      "SNP 827\n",
      "SNP 828\n",
      "SNP 829\n",
      "SNP 830\n",
      "SNP 831\n",
      "SNP 832\n",
      "SNP 833\n",
      "SNP 834\n",
      "SNP 835\n",
      "SNP 836\n",
      "SNP 837\n",
      "SNP 838\n",
      "SNP 839\n",
      "SNP 840\n",
      "SNP 841\n",
      "SNP 842\n",
      "SNP 843\n",
      "SNP 844\n",
      "SNP 845\n",
      "SNP 846\n",
      "SNP 847\n",
      "SNP 848\n",
      "SNP 849\n",
      "SNP 850\n",
      "SNP 851\n",
      "SNP 852\n",
      "SNP 853\n",
      "SNP 854\n",
      "SNP 855\n",
      "SNP 856\n",
      "SNP 857\n",
      "SNP 858\n",
      "SNP 859\n",
      "SNP 860\n",
      "SNP 861\n",
      "SNP 862\n",
      "SNP 863\n",
      "SNP 864\n",
      "SNP 865\n",
      "SNP 866\n",
      "SNP 867\n",
      "SNP 868\n",
      "SNP 869\n",
      "SNP 870\n",
      "SNP 871\n",
      "SNP 872\n",
      "SNP 873\n",
      "SNP 874\n",
      "SNP 875\n",
      "SNP 876\n",
      "SNP 877\n",
      "SNP 878\n",
      "SNP 879\n",
      "SNP 880\n",
      "SNP 881\n",
      "SNP 882\n",
      "SNP 883\n",
      "SNP 884\n",
      "SNP 885\n",
      "SNP 886\n",
      "SNP 887\n",
      "SNP 888\n",
      "SNP 889\n",
      "SNP 890\n",
      "SNP 891\n",
      "SNP 892\n",
      "SNP 893\n",
      "SNP 894\n",
      "SNP 895\n",
      "SNP 896\n",
      "SNP 897\n",
      "SNP 898\n",
      "SNP 899\n",
      "SNP 900\n",
      "SNP 901\n",
      "SNP 902\n",
      "SNP 903\n",
      "SNP 904\n",
      "SNP 905\n",
      "SNP 906\n",
      "SNP 907\n",
      "SNP 908\n",
      "SNP 909\n",
      "SNP 910\n",
      "SNP 911\n",
      "SNP 912\n",
      "SNP 913\n",
      "SNP 914\n",
      "SNP 915\n",
      "SNP 916\n",
      "SNP 917\n",
      "SNP 918\n",
      "SNP 919\n",
      "SNP 920\n",
      "SNP 921\n",
      "SNP 922\n",
      "SNP 923\n",
      "SNP 924\n",
      "SNP 925\n",
      "SNP 926\n",
      "SNP 927\n",
      "SNP 928\n",
      "SNP 929\n",
      "SNP 930\n",
      "SNP 931\n",
      "SNP 932\n",
      "SNP 933\n",
      "SNP 934\n",
      "SNP 935\n",
      "SNP 936\n",
      "SNP 937\n",
      "SNP 938\n",
      "SNP 939\n",
      "SNP 940\n",
      "SNP 941\n",
      "SNP 942\n",
      "SNP 943\n",
      "SNP 944\n",
      "SNP 945\n",
      "SNP 946\n",
      "SNP 947\n",
      "SNP 948\n",
      "SNP 949\n",
      "SNP 950\n",
      "SNP 951\n",
      "SNP 952\n",
      "SNP 953\n",
      "SNP 954\n",
      "SNP 955\n",
      "SNP 956\n",
      "SNP 957\n",
      "SNP 958\n",
      "SNP 959\n",
      "SNP 960\n",
      "SNP 961\n",
      "SNP 962\n",
      "SNP 963\n",
      "SNP 964\n",
      "SNP 965\n",
      "SNP 966\n",
      "SNP 967\n",
      "SNP 968\n",
      "SNP 969\n",
      "SNP 970\n",
      "SNP 971\n",
      "SNP 972\n",
      "SNP 973\n",
      "SNP 974\n",
      "SNP 975\n",
      "SNP 976\n",
      "SNP 977\n",
      "SNP 978\n",
      "SNP 979\n",
      "SNP 980\n",
      "SNP 981\n",
      "SNP 982\n",
      "SNP 983\n",
      "SNP 984\n",
      "SNP 985\n",
      "SNP 986\n",
      "SNP 987\n",
      "SNP 988\n",
      "SNP 989\n",
      "SNP 990\n",
      "SNP 991\n",
      "SNP 992\n",
      "SNP 993\n",
      "SNP 994\n",
      "SNP 995\n",
      "SNP 996\n",
      "SNP 997\n",
      "SNP 998\n",
      "SNP 999\n",
      "Chromosome 2\n",
      "SNP 0\n",
      "SNP 1\n",
      "SNP 2\n",
      "SNP 3\n",
      "SNP 4\n",
      "SNP 5\n",
      "SNP 6\n",
      "SNP 7\n",
      "SNP 8\n",
      "SNP 9\n",
      "SNP 10\n",
      "SNP 11\n",
      "SNP 12\n",
      "SNP 13\n",
      "SNP 14\n",
      "SNP 15\n",
      "SNP 16\n",
      "SNP 17\n",
      "SNP 18\n",
      "SNP 19\n",
      "SNP 20\n",
      "SNP 21\n",
      "SNP 22\n",
      "SNP 23\n",
      "SNP 24\n",
      "SNP 25\n",
      "SNP 26\n",
      "SNP 27\n",
      "SNP 28\n",
      "SNP 29\n",
      "SNP 30\n",
      "SNP 31\n",
      "SNP 32\n",
      "SNP 33\n",
      "SNP 34\n",
      "SNP 35\n",
      "SNP 36\n",
      "SNP 37\n",
      "SNP 38\n",
      "SNP 39\n",
      "SNP 40\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNP 41\n",
      "SNP 42\n",
      "SNP 43\n",
      "SNP 44\n",
      "SNP 45\n",
      "SNP 46\n",
      "SNP 47\n",
      "SNP 48\n",
      "SNP 49\n",
      "SNP 50\n",
      "SNP 51\n",
      "SNP 52\n",
      "SNP 53\n",
      "SNP 54\n",
      "SNP 55\n",
      "SNP 56\n",
      "SNP 57\n",
      "SNP 58\n",
      "SNP 59\n",
      "SNP 60\n",
      "SNP 61\n",
      "SNP 62\n",
      "SNP 63\n",
      "SNP 64\n",
      "SNP 65\n",
      "SNP 66\n",
      "SNP 67\n",
      "SNP 68\n",
      "SNP 69\n",
      "SNP 70\n",
      "SNP 71\n",
      "SNP 72\n",
      "SNP 73\n",
      "SNP 74\n",
      "SNP 75\n",
      "SNP 76\n",
      "SNP 77\n",
      "SNP 78\n",
      "SNP 79\n",
      "SNP 80\n",
      "SNP 81\n",
      "SNP 82\n",
      "SNP 83\n",
      "SNP 84\n",
      "SNP 85\n",
      "SNP 86\n",
      "SNP 87\n",
      "SNP 88\n",
      "SNP 89\n",
      "SNP 90\n",
      "SNP 91\n",
      "SNP 92\n",
      "SNP 93\n",
      "SNP 94\n",
      "SNP 95\n",
      "SNP 96\n",
      "SNP 97\n",
      "SNP 98\n",
      "SNP 99\n",
      "SNP 100\n",
      "SNP 101\n",
      "SNP 102\n",
      "SNP 103\n",
      "SNP 104\n",
      "SNP 105\n",
      "SNP 106\n",
      "SNP 107\n",
      "SNP 108\n",
      "SNP 109\n",
      "SNP 110\n",
      "SNP 111\n",
      "SNP 112\n",
      "SNP 113\n",
      "SNP 114\n",
      "SNP 115\n",
      "SNP 116\n",
      "SNP 117\n",
      "SNP 118\n",
      "SNP 119\n",
      "SNP 120\n",
      "SNP 121\n",
      "SNP 122\n",
      "SNP 123\n",
      "SNP 124\n",
      "SNP 125\n",
      "SNP 126\n",
      "SNP 127\n",
      "SNP 128\n",
      "SNP 129\n",
      "SNP 130\n",
      "SNP 131\n",
      "SNP 132\n",
      "SNP 133\n",
      "SNP 134\n",
      "SNP 135\n",
      "SNP 136\n",
      "SNP 137\n",
      "SNP 138\n",
      "SNP 139\n",
      "SNP 140\n",
      "SNP 141\n",
      "SNP 142\n",
      "SNP 143\n",
      "SNP 144\n",
      "SNP 145\n",
      "SNP 146\n",
      "SNP 147\n",
      "SNP 148\n",
      "SNP 149\n",
      "SNP 150\n",
      "SNP 151\n",
      "SNP 152\n",
      "SNP 153\n",
      "SNP 154\n",
      "SNP 155\n",
      "SNP 156\n",
      "SNP 157\n",
      "SNP 158\n",
      "SNP 159\n",
      "SNP 160\n",
      "SNP 161\n",
      "SNP 162\n",
      "SNP 163\n",
      "SNP 164\n",
      "SNP 165\n",
      "SNP 166\n",
      "SNP 167\n",
      "SNP 168\n",
      "SNP 169\n",
      "SNP 170\n",
      "SNP 171\n",
      "SNP 172\n",
      "SNP 173\n",
      "SNP 174\n",
      "SNP 175\n",
      "SNP 176\n",
      "SNP 177\n",
      "SNP 178\n",
      "SNP 179\n",
      "SNP 180\n",
      "SNP 181\n",
      "SNP 182\n",
      "SNP 183\n",
      "SNP 184\n",
      "SNP 185\n",
      "SNP 186\n",
      "SNP 187\n",
      "SNP 188\n",
      "SNP 189\n",
      "SNP 190\n",
      "SNP 191\n",
      "SNP 192\n",
      "SNP 193\n",
      "SNP 194\n",
      "SNP 195\n",
      "SNP 196\n",
      "SNP 197\n",
      "SNP 198\n",
      "SNP 199\n",
      "SNP 200\n",
      "SNP 201\n",
      "SNP 202\n",
      "SNP 203\n",
      "SNP 204\n",
      "SNP 205\n",
      "SNP 206\n",
      "SNP 207\n",
      "SNP 208\n",
      "SNP 209\n",
      "SNP 210\n",
      "SNP 211\n",
      "SNP 212\n",
      "SNP 213\n",
      "SNP 214\n",
      "SNP 215\n",
      "SNP 216\n",
      "SNP 217\n",
      "SNP 218\n",
      "SNP 219\n",
      "SNP 220\n",
      "SNP 221\n",
      "SNP 222\n",
      "SNP 223\n",
      "SNP 224\n",
      "SNP 225\n",
      "SNP 226\n",
      "SNP 227\n",
      "SNP 228\n",
      "SNP 229\n",
      "SNP 230\n",
      "SNP 231\n",
      "SNP 232\n",
      "SNP 233\n",
      "SNP 234\n",
      "SNP 235\n",
      "SNP 236\n",
      "SNP 237\n",
      "SNP 238\n",
      "SNP 239\n",
      "SNP 240\n",
      "SNP 241\n",
      "SNP 242\n",
      "SNP 243\n",
      "SNP 244\n",
      "SNP 245\n",
      "SNP 246\n",
      "SNP 247\n",
      "SNP 248\n",
      "SNP 249\n",
      "SNP 250\n",
      "SNP 251\n",
      "SNP 252\n",
      "SNP 253\n",
      "SNP 254\n",
      "SNP 255\n",
      "SNP 256\n",
      "SNP 257\n",
      "SNP 258\n",
      "SNP 259\n",
      "SNP 260\n",
      "SNP 261\n",
      "SNP 262\n",
      "SNP 263\n",
      "SNP 264\n",
      "SNP 265\n",
      "SNP 266\n",
      "SNP 267\n",
      "SNP 268\n",
      "SNP 269\n",
      "SNP 270\n",
      "SNP 271\n",
      "SNP 272\n",
      "SNP 273\n",
      "SNP 274\n",
      "SNP 275\n",
      "SNP 276\n",
      "SNP 277\n",
      "SNP 278\n",
      "SNP 279\n",
      "SNP 280\n",
      "SNP 281\n",
      "SNP 282\n",
      "SNP 283\n",
      "SNP 284\n",
      "SNP 285\n",
      "SNP 286\n",
      "SNP 287\n",
      "SNP 288\n",
      "SNP 289\n",
      "SNP 290\n",
      "SNP 291\n",
      "SNP 292\n",
      "SNP 293\n",
      "SNP 294\n",
      "SNP 295\n",
      "SNP 296\n",
      "SNP 297\n",
      "SNP 298\n",
      "SNP 299\n",
      "SNP 300\n",
      "SNP 301\n",
      "SNP 302\n",
      "SNP 303\n",
      "SNP 304\n",
      "SNP 305\n",
      "SNP 306\n",
      "SNP 307\n",
      "SNP 308\n",
      "SNP 309\n",
      "SNP 310\n",
      "SNP 311\n",
      "SNP 312\n",
      "SNP 313\n",
      "SNP 314\n",
      "SNP 315\n",
      "SNP 316\n",
      "SNP 317\n",
      "SNP 318\n",
      "SNP 319\n",
      "SNP 320\n",
      "SNP 321\n"
     ]
    }
   ],
   "source": [
    "def runGWAS (disease):\n",
    "    \"\"\"\n",
    "    1. For each chromosome 1:22 call the gwas function\n",
    "    2. Merge all the result files into one \n",
    "    3. Filter your data\n",
    "       a. remove all SNPs that deviate signficantly from HWE equilibrium -- these may indicate an error or \n",
    "          population specific deviation. For this remove all SNPs for which HWE deviation P-Value < 0.05\n",
    "       b. remove all SNPs that have minor allele frequency less than 1% -- not enough data\n",
    "    4. Draw a manhattan plot\n",
    "       c. draw a line at your Bonferroni threshold (Bonferroni correction): \n",
    "       if your significance is set at p.value < 0.05, then Bonferroni correction = 0.05/number of tests that you ran\n",
    "    \"\"\"\n",
    "    import os\n",
    "    \n",
    "    dataName = \"LittleBoy\"\n",
    "    path = \"\"\n",
    "    if(dataName=='Japan'):\n",
    "        path = \"/Users/varun/Desktop/BeaverWorks/Week3Public/BWSI_100set/\"\n",
    "    elif(dataName=='LittleBoy'):\n",
    "        path = \"/Users/varun/Desktop/BeaverWorks/Week3Public/BWSI_100set/LittleBoy\"\n",
    "    elif(dataName=='FatMan'):\n",
    "        path = \"/Users/varun/Desktop/BeaverWorks/Week3Public/BWSI_100set/FatMan\"\n",
    "    \n",
    "    data = pd.DataFrame(columns = [\"Chromosome\",\"SNP\",\"RSID\", \"Minor Allele\",\"Major Allele\",\"Minor Allele Control Frequency\",\"Major Allele Control Frequency\",\"Minor Allele Disease Frequency\",\"Major Allele Disease Frequency\",\"Odds Ratio\",\"Odds Ratio p-value\",\"HWE\", \"-log(Odds Ratio p-value)\", \"-log(HWE)\"])\n",
    "    index = 0\n",
    "    for chrom_num in range(1, 23):\n",
    "        print(\"Chromosome \" + str(chrom_num))\n",
    "        \n",
    "        if(chrom_num<10):\n",
    "            fileName = path + '/' + str(disease) + '/Affx_gt_' + str(disease) + '_Chiamo_0' + str(chrom_num) + '.tped.gz'   \n",
    "        else:\n",
    "            fileName = path + '/' + str(disease) + '/Affx_gt_' + str(disease) + '_Chiamo_' + str(chrom_num) + '.tped.gz'\n",
    "        data = pd.read_csv(fileName, compression='gzip', header=None, sep='\\t')\n",
    "        \n",
    "        if chrom_num < 10:\n",
    "            chrom_num = \"0\"+str(chrom_num)\n",
    "        control_file = path + '/58C/Affx_gt_58C_Chiamo_' + str(chrom_num) + '.tped.gz'\n",
    "        control_58c = pd.read_csv(control_file, compression='gzip', header=None, sep='\\t')\n",
    "        \n",
    "        \n",
    "        control_58c_header = control_58c.loc[:, :3]\n",
    "        control_58c_body = control_58c.loc[:, 4:]\n",
    "\n",
    "        ## Load the test data \n",
    "        #     t1d_22_file = '/Users/varun/Desktop/BeaverWorks/Week3Public/BWSI_100set/T1D/Affx_gt_T2D_Chiamo_22.tped.gz'\n",
    "        t1d = data\n",
    "\n",
    "        t1d_header = data.loc[:, :3]\n",
    "        t1d_body = data.loc[:,4:]\n",
    "\n",
    "        snp_file = path + '/58C/snps/snps_'+str(chrom_num)\n",
    "        snp = pd.read_csv(snp_file, header=None, sep='\\t')\n",
    "        \n",
    "    \n",
    "        \n",
    "        \n",
    "        combined = pd.concat([control_58c_header, snp], axis=1)\n",
    "        for j in range(1000):\n",
    "            x = gwas(disease, chrom_num, j, t1d, control_58c_body, combined)\n",
    "            print(\"SNP \" + str(j))\n",
    "            if(x.empty!=True):\n",
    "                data.loc[index] = x\n",
    "                index += 1\n",
    "        data.to_csv(\"data_long.csv\", sep='\\t')\n",
    "            \n",
    "\n",
    "    \n",
    "    \n",
    "#     df = pd.DataFrame({\"Chromosome\":arr[0],\n",
    "#                   \"SNP\": arr[1],\n",
    "#                    \"Minor Allele\": arr[9],\n",
    "#                    \"Major Allele\": arr[10],\n",
    "#                   \"Minor Allele Control Frequency\": arr[2],\n",
    "#                   \"Major Allele Control Frequency\": arr[3],\n",
    "#                   \"Minor Allele Disease Frequency\": arr[4],\n",
    "#                   \"Major Allele Disease Frequency\": arr[5],\n",
    "#                   \"Odds Ratio\": arr[6],\n",
    "#                   \"Odds Ratio p-value\": arr[7],\n",
    "#                   \"HWE\": arr[8]})\n",
    "    \n",
    "    data.to_csv(\"data_long.csv\", sep='\\t')\n",
    "\n",
    "runGWAS('T1D')\n",
    "print(\"Done!\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src=\"manhattan.png\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## How many SNPs did you identify that exceeded the Bonferroni-level of significance?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## How much can you trust your results\n",
    "\n",
    "Typically, you want to examine the genome-wide distribution of your chi_square distribution against an expected null distribution. Visually you can do this using a quantile-quantile plot. Significant deviation from the null distribution may point to population stratification, familial relationships, technical bias, poor sample collection or sometimes even undetected sample duplications.\n",
    "(Slide courtesy - Dr. Chirag Patel)\n",
    "\n",
    "<img src=\"qqplot_example.png\">\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Based on your results do a quantile-quantile plot for your data\n",
    "\n",
    "##load your results \n",
    "##sort/order the -log10 of your p values in increasing order\n",
    "##generate an expected set of -log10(p_values)\n",
    "##plot the expected values on the x-axis and the observed -log10 (pvalues) on the y axis\n",
    "##Alternatively use: p = observed p values\n",
    "##stats.probplot(p, dist=\"norm\", plot=pylab)\n",
    "##pylab.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Genomic Inflation Factor\n",
    "\n",
    "Mathematically, you can correct for some of the deviation by computing an inflation factor and then correcting for this inflation factor. This quantity is termed as genomic inflation factor ($\\lambda$)\n",
    "\n",
    "$\\lambda$ = observed median of test statistic distribution / expected median of the test statistic \n",
    "distribution.  For 1 degree of freedom, the X2 distribution has an expected median of 0.455\n",
    "\n",
    "$\\lambda$ <= 1.05 is considered acceptable. >1.1 is troubling, and indicates there is some inflation of the p values.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Compute the genomic inflation factor\n",
    "\n",
    "##labmda =  median(CHI2)/qchisq(0.5, 1)\n",
    "##if (gif > 1.05), correct your computed Chi.square values and recompute the p_values\n",
    "##chi2_corrected = CHI2/lambda, \n",
    "##pval_corrected = pchisq(chi2_corrected, 1, lower.tail = F)\n",
    "##re-draw the qqplot using the corrected pvalues (pval_corrected)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## What does this mean?\n",
    "\n",
    "Identify a SNP that is significant (above the Bonferroni threshold). \n",
    "What is the p-value of association and odds ratio for the SNP?\n",
    "What is the interpretation of the odds ratio? \n",
    "What gene is this SNP associated with and its putative function in your disease pathogenesis? \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Polygenic Risk Scores (PRS)\n",
    "\n",
    "A PRS is the cumulative risk of all SNP for a single patient. For example, if a patient has a 2 snps with OR=2 each, then their PRS is 4. The underlying assumption here is that each allele has an equal contribution and there is no interation between SNPS, that is the SNPS are independent. \n",
    "\n",
    "$prs =\\sum_{i=1}^{m}-log_{10}(x_i) \\cdot n_i$"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
