{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Challenge Problem 3: Genome Wide Association Study"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Background:\n",
    "\n",
    "(Adapted From BMI 704, Dr. Chirag Patel)\n",
    "\n",
    "Genome-wide association studies (GWASs) are arguably the most significant biomedical advance in the last decade. Utilizing an observational case-control design, populations with disease are analytically compared to those without to discern patterns in differences in genetic variant, or single nucleotide polymorphism (SNP),frequencies in each of the respective populations. If statistically differences in SNP frequencies in cases vs\n",
    "controls are found, one can conclude the SNP locus may have a causal relationship, implicating a gene, genetic region, or gene regulator in the disease of interest.\n",
    "\n",
    "Analyzing GWASs provide lessons in observational research, including statistical association and multiple testing. They also provide a way to create hypotheses about biological basis of disease for further study. We will execute a GWAS analysis using data from the Wellcome Trust Case-Control Consortium (WTCCC) on Type 2 Diabetes, Type 1 Diabetes, Bipolar Disorder, Rheumatoid Arthritis, Chron's Disease and Hypertension versus 1,500 controls.\n",
    "\n",
    "Note: There are standard tools and pipelines available to do this but we will do it the old fashioned way!\n",
    "\n",
    "The Data is provided in separate directories under the primary WTCCC directory. The subdirectories are labeled as such:\n",
    "* 58C : control population 1\n",
    "* NBS : control population 2\n",
    "* RA : rheumatoid arthritis\n",
    "* T2D : type 2 diabetes\n",
    "* T1D : type 1 diabetes\n",
    "* CAD : coronary artery disease\n",
    "* HT : hypertension\n",
    "* CD : Crohn’s disease\n",
    "* BD : bipolar disorder\n",
    "\n",
    "All of the GWAS data is broken down per chromosome, 1-22 plus X and is written into the file name, as well as the disease. For example:\n",
    "\n",
    "Affx_gt_ 58C _Chiamo_ 06 .tped.gz\n",
    "\n",
    "Contains GWAS data for the 6th (06) chromosome on control population 1 (58C). The biggest files are chromosome 1 and 2 and smallest is 22.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Each row is a SNP (there are 6207 SNPs measured in Chromosome 22). The first through fourth columns denote the chromosome number, the second is the WTCCC snp_id (not to be confused with the rsid), the start coordinate (0), and the end coordinate. Thereafter, each column is a genotype (where each base is separated by a space character) for that SNP for each individual.\n",
    "\n",
    "Additionally, each subdirectory also contains a ‘snps_info.tar.gz’, which contains the snp_id and rsID for each SNP for each chromosome. These files can enable you to map the WTCCC snp_id to the rsid, which is helpful to find the closest genes to a SNP (for example when you query the GWAS catalog)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/varun/Desktop/BeaverWorks/Week3Public/GWAS\r\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from os import path\n",
    "import csv\n",
    "import numpy as np\n",
    "from time import time\n",
    "!pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Load the test data \n",
    "t1d_22_file = '/Users/varun/Desktop/BeaverWorks/Week3Public/BWSI_100set/T2D/Affx_gt_T2D_Chiamo_22.tped.gz'\n",
    "t1d = pd.read_csv(t1d_22_file, compression='gzip', header=None, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(100, 2003)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Look at the shape of the data frame - you should have 6207 SNPs for 2003 individuals\n",
    "t1d.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22</td>\n",
       "      <td>SNP_A-4234155</td>\n",
       "      <td>0</td>\n",
       "      <td>32056663</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>22</td>\n",
       "      <td>SNP_A-2223191</td>\n",
       "      <td>0</td>\n",
       "      <td>35653934</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>22</td>\n",
       "      <td>SNP_A-4202732</td>\n",
       "      <td>0</td>\n",
       "      <td>32956968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>22</td>\n",
       "      <td>SNP_A-1811354</td>\n",
       "      <td>0</td>\n",
       "      <td>46997362</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>22</td>\n",
       "      <td>SNP_A-1797179</td>\n",
       "      <td>0</td>\n",
       "      <td>32955711</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    0              1  2         3\n",
       "0  22  SNP_A-4234155  0  32056663\n",
       "1  22  SNP_A-2223191  0  35653934\n",
       "2  22  SNP_A-4202732  0  32956968\n",
       "3  22  SNP_A-1811354  0  46997362\n",
       "4  22  SNP_A-1797179  0  32955711"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Look at the top 10 rows\n",
    "# t2d.head(10)\n",
    "t1d_header = t1d.loc[:, :3]\n",
    "t1d_body = t1d.loc[:,4:]\n",
    "t1d_header.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>1498</th>\n",
       "      <th>1499</th>\n",
       "      <th>1500</th>\n",
       "      <th>1501</th>\n",
       "      <th>1502</th>\n",
       "      <th>1503</th>\n",
       "      <th>1504</th>\n",
       "      <th>1505</th>\n",
       "      <th>1506</th>\n",
       "      <th>1507</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22</td>\n",
       "      <td>SNP_A-4234155</td>\n",
       "      <td>0</td>\n",
       "      <td>32056663</td>\n",
       "      <td>C T</td>\n",
       "      <td>C T</td>\n",
       "      <td>C T</td>\n",
       "      <td>T T</td>\n",
       "      <td>C T</td>\n",
       "      <td>T T</td>\n",
       "      <td>...</td>\n",
       "      <td>T T</td>\n",
       "      <td>C C</td>\n",
       "      <td>C T</td>\n",
       "      <td>C T</td>\n",
       "      <td>C T</td>\n",
       "      <td>C C</td>\n",
       "      <td>C T</td>\n",
       "      <td>T T</td>\n",
       "      <td>C T</td>\n",
       "      <td>T T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>22</td>\n",
       "      <td>SNP_A-2223191</td>\n",
       "      <td>0</td>\n",
       "      <td>35653934</td>\n",
       "      <td>T T</td>\n",
       "      <td>C C</td>\n",
       "      <td>C T</td>\n",
       "      <td>T T</td>\n",
       "      <td>C C</td>\n",
       "      <td>C T</td>\n",
       "      <td>...</td>\n",
       "      <td>C T</td>\n",
       "      <td>C T</td>\n",
       "      <td>C T</td>\n",
       "      <td>C T</td>\n",
       "      <td>T T</td>\n",
       "      <td>C T</td>\n",
       "      <td>C T</td>\n",
       "      <td>C C</td>\n",
       "      <td>T T</td>\n",
       "      <td>C C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>22</td>\n",
       "      <td>SNP_A-4202732</td>\n",
       "      <td>0</td>\n",
       "      <td>32956968</td>\n",
       "      <td>G G</td>\n",
       "      <td>A G</td>\n",
       "      <td>A G</td>\n",
       "      <td>G G</td>\n",
       "      <td>A G</td>\n",
       "      <td>A G</td>\n",
       "      <td>...</td>\n",
       "      <td>A G</td>\n",
       "      <td>A G</td>\n",
       "      <td>A G</td>\n",
       "      <td>G G</td>\n",
       "      <td>G G</td>\n",
       "      <td>G G</td>\n",
       "      <td>G G</td>\n",
       "      <td>G G</td>\n",
       "      <td>A G</td>\n",
       "      <td>G G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>22</td>\n",
       "      <td>SNP_A-1811354</td>\n",
       "      <td>0</td>\n",
       "      <td>46997362</td>\n",
       "      <td>A G</td>\n",
       "      <td>A A</td>\n",
       "      <td>A G</td>\n",
       "      <td>A A</td>\n",
       "      <td>A G</td>\n",
       "      <td>G G</td>\n",
       "      <td>...</td>\n",
       "      <td>A A</td>\n",
       "      <td>A G</td>\n",
       "      <td>A A</td>\n",
       "      <td>A A</td>\n",
       "      <td>A G</td>\n",
       "      <td>A G</td>\n",
       "      <td>A A</td>\n",
       "      <td>A A</td>\n",
       "      <td>A G</td>\n",
       "      <td>A A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>22</td>\n",
       "      <td>SNP_A-1797179</td>\n",
       "      <td>0</td>\n",
       "      <td>32955711</td>\n",
       "      <td>G G</td>\n",
       "      <td>A G</td>\n",
       "      <td>A G</td>\n",
       "      <td>G G</td>\n",
       "      <td>A G</td>\n",
       "      <td>A G</td>\n",
       "      <td>...</td>\n",
       "      <td>A G</td>\n",
       "      <td>A G</td>\n",
       "      <td>A G</td>\n",
       "      <td>G G</td>\n",
       "      <td>G G</td>\n",
       "      <td>G G</td>\n",
       "      <td>G G</td>\n",
       "      <td>G G</td>\n",
       "      <td>A G</td>\n",
       "      <td>G G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 1508 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   0              1     2         3    4    5    6    7    8    9     ...  \\\n",
       "0    22  SNP_A-4234155     0  32056663  C T  C T  C T  T T  C T  T T  ...   \n",
       "1    22  SNP_A-2223191     0  35653934  T T  C C  C T  T T  C C  C T  ...   \n",
       "2    22  SNP_A-4202732     0  32956968  G G  A G  A G  G G  A G  A G  ...   \n",
       "3    22  SNP_A-1811354     0  46997362  A G  A A  A G  A A  A G  G G  ...   \n",
       "4    22  SNP_A-1797179     0  32955711  G G  A G  A G  G G  A G  A G  ...   \n",
       "\n",
       "  1498 1499 1500 1501 1502 1503 1504 1505 1506 1507  \n",
       "0  T T  C C  C T  C T  C T  C C  C T  T T  C T  T T  \n",
       "1  C T  C T  C T  C T  T T  C T  C T  C C  T T  C C  \n",
       "2  A G  A G  A G  G G  G G  G G  G G  G G  A G  G G  \n",
       "3  A A  A G  A A  A A  A G  A G  A A  A A  A G  A A  \n",
       "4  A G  A G  A G  G G  G G  G G  G G  G G  A G  G G  \n",
       "\n",
       "[5 rows x 1508 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "control_file = '/Users/varun/Desktop/BeaverWorks/Week3Public/BWSI_100set/58C/Affx_gt_58C_Chiamo_22.tped.gz'\n",
    "control_58c = pd.read_csv(control_file, compression='gzip', header=None, sep='\\t')\n",
    "\n",
    "control_58c_header = control_58c.loc[:, :3]\n",
    "control_58c_body = control_58c.loc[:, 4:]\n",
    "control_58c.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# control_file = '/Users/varun/Desktop/BeaverWorks/Week3Public/BWSI_100set/NBS/Affx_gt_NBS_Chiamo_01.tped.gz'\n",
    "# control_nbs = pd.read_csv(control_file, compression='gzip', header=None, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# control_nbs_header = control_nbs.loc[:, :3]\n",
    "# control_nbs_body = control_nbs.loc[:, 4:]\n",
    "# control_nbs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>SNP</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22</td>\n",
       "      <td>SNP_A-4234155</td>\n",
       "      <td>0</td>\n",
       "      <td>32056663</td>\n",
       "      <td>rs12484156</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>22</td>\n",
       "      <td>SNP_A-2223191</td>\n",
       "      <td>0</td>\n",
       "      <td>35653934</td>\n",
       "      <td>rs909486</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>22</td>\n",
       "      <td>SNP_A-4202732</td>\n",
       "      <td>0</td>\n",
       "      <td>32956968</td>\n",
       "      <td>rs1873230</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>22</td>\n",
       "      <td>SNP_A-1811354</td>\n",
       "      <td>0</td>\n",
       "      <td>46997362</td>\n",
       "      <td>rs7287281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>22</td>\n",
       "      <td>SNP_A-1797179</td>\n",
       "      <td>0</td>\n",
       "      <td>32955711</td>\n",
       "      <td>rs1003935</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    0            SNP  2         3           4\n",
       "0  22  SNP_A-4234155  0  32056663  rs12484156\n",
       "1  22  SNP_A-2223191  0  35653934    rs909486\n",
       "2  22  SNP_A-4202732  0  32956968   rs1873230\n",
       "3  22  SNP_A-1811354  0  46997362   rs7287281\n",
       "4  22  SNP_A-1797179  0  32955711   rs1003935"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp_file = '/Users/varun/Desktop/BeaverWorks/Week3Public/BWSI_100set/58C/snps/snps_22'\n",
    "snp = pd.read_csv(snp_file, header=None, sep='\\t')\n",
    "\n",
    "combined = pd.concat([control_58c_header, snp], axis=1)\n",
    "# combined = combined[:100]\n",
    "\n",
    "control_58c_header = control_58c_header.rename(columns={1:'SNP'})\n",
    "snp = snp.loc[:, 3:]\n",
    "snp = snp.rename(columns={3:'SNP'})\n",
    "combined = control_58c_header.join(snp.set_index('SNP'), on='SNP')\n",
    "combined.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>SNP</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22</td>\n",
       "      <td>SNP_A-4234155</td>\n",
       "      <td>0</td>\n",
       "      <td>32056663</td>\n",
       "      <td>rs12484156</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>22</td>\n",
       "      <td>SNP_A-2223191</td>\n",
       "      <td>0</td>\n",
       "      <td>35653934</td>\n",
       "      <td>rs909486</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>22</td>\n",
       "      <td>SNP_A-4202732</td>\n",
       "      <td>0</td>\n",
       "      <td>32956968</td>\n",
       "      <td>rs1873230</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>22</td>\n",
       "      <td>SNP_A-1811354</td>\n",
       "      <td>0</td>\n",
       "      <td>46997362</td>\n",
       "      <td>rs7287281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>22</td>\n",
       "      <td>SNP_A-1797179</td>\n",
       "      <td>0</td>\n",
       "      <td>32955711</td>\n",
       "      <td>rs1003935</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    0            SNP  2         3           4\n",
       "0  22  SNP_A-4234155  0  32056663  rs12484156\n",
       "1  22  SNP_A-2223191  0  35653934    rs909486\n",
       "2  22  SNP_A-4202732  0  32956968   rs1873230\n",
       "3  22  SNP_A-1811354  0  46997362   rs7287281\n",
       "4  22  SNP_A-1797179  0  32955711   rs1003935"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined = pd.concat([t1d_header, snp], axis=1)\n",
    "# combined = combined[:100]\n",
    "\n",
    "t1d_header = t1d_header.rename(columns={1:'SNP'})\n",
    "combined_t1d = t1d_header.join(snp.set_index('SNP'), on='SNP')\n",
    "combined_t1d.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# row = 0\n",
    "# t1d_1 = t1d.loc[row, 4:]\n",
    "# # print(t1d)\n",
    "\n",
    "# control_58c_1 = control_58c_body.loc[0]\n",
    "# print(\"Snip: \" + str(t1d.loc[row, 1])[6:])\n",
    "# print(\"RSID: \" + str(combined_t1d.loc[row, 4])[2:])\n",
    "# print(\"Type 2 Diabetes:\")\n",
    "# print(computeAlleleFrequency(getAlleleCounts(t1d_1)[0], getAlleleCounts(t1d_1)[1]))\n",
    "# t1_minor = computeAlleleFrequency(getAlleleCounts(t1d_1)[0], getAlleleCounts(t1d_1)[1])\n",
    "# print(1-computeAlleleFrequency(getAlleleCounts(t1d_1)[0], getAlleleCounts(t1d_1)[1]))\n",
    "# t1_major = 1 - t1_minor\n",
    "# print(\"\\nControl:\")\n",
    "# print(computeAlleleFrequency(getAlleleCounts(control_58c_1)[0], getAlleleCounts(control_58c_1)[1]))\n",
    "# control_minor = computeAlleleFrequency(getAlleleCounts(control_58c_1)[0], getAlleleCounts(control_58c_1)[1])\n",
    "# print(1-computeAlleleFrequency(getAlleleCounts(control_58c_1)[0], getAlleleCounts(control_58c_1)[1]))\n",
    "# control_major = 1-control_minor\n",
    "# # print(control)\n",
    "# print(\"\\nOdds Ratio:\")\n",
    "# print(computeOR([t1_minor, t1_major, control_minor, control_major]))\n",
    "\n",
    "# print(\"\\nChi-Squared:\")\n",
    "# print(getPValue([t1_minor, t1_major]))\n",
    "# print(getPValue([control_minor, control_major]))\n",
    "\n",
    "# print(\"\\nHWE:\")\n",
    "# print(HWEChiSq(t1_minor, [t1d_1.to_string().count('C C'), t1d_1.to_string().count('C T'), t1d_1.to_string().count('T T')]))\n",
    "# print(HWEChiSq(control_minor, [control_58c_1.to_string().count('C C'), control_58c_1.to_string().count('C T'), control_58c_1.to_string().count('T T')]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def findAlleles(genotype):\n",
    "    arr = []\n",
    "    shit = []\n",
    "    gen = genotype.str\n",
    "    c = gen.count('C').sum()\n",
    "    t = gen.count('T').sum()\n",
    "    a = gen.count('A').sum()\n",
    "    g = gen.count('G').sum()\n",
    "    \n",
    "    if c>0:\n",
    "        arr.append('C')\n",
    "        shit.append(c)\n",
    "    if t>0:\n",
    "        arr.append('T')\n",
    "        shit.append(t)\n",
    "    if a>0:\n",
    "        arr.append('A')\n",
    "        shit.append(a)\n",
    "    if g>0:\n",
    "        arr.append('G')\n",
    "        shit.append(g)\n",
    "        \n",
    "    if len(arr) == 1:\n",
    "        return None\n",
    "    elif shit[0] > shit[1]:\n",
    "        return [arr[1], arr[0]]\n",
    "    else:\n",
    "        return arr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "## function to load the data file\n",
    "def loadFile (disease, chrom_num):\n",
    "    ## construct your file name based on the disease and chrom_number\n",
    "    if(chrom_num<10):\n",
    "        fileName = '/Users/varun/Desktop/BeaverWorks/Week3Public/BWSI_100set/' + str(disease) + '/Affx_gt_' + str(disease) + '_Chiamo_0' + str(chrom_num) + '.tped.gz'   \n",
    "    else:\n",
    "        fileName = '/Users/varun/Desktop/BeaverWorks/Week3Public/BWSI_100set/' + str(disease) + '/Affx_gt_' + str(disease) + '_Chiamo_' + str(chrom_num) + '.tped.gz'\n",
    "    dat = pd.read_csv(fileName, compression='gzip', header=None, sep='\\t')\n",
    "    return (dat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "## compute Allele Frequency for the minor allele\n",
    "def computeAlleleFrequency (f_C, f_T):\n",
    "    \n",
    "    \"\"\"\n",
    "    f_C = minor allele count\n",
    "    f_T = major allele count\n",
    "    minor_allele_frequency = f_C/ (f_C+f_T) \n",
    "    \"\"\"\n",
    "    minor_allele_frequency = f_C/ (f_C+f_T) \n",
    "    return (minor_allele_frequency)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getAlleleCounts (genotype, arr):\n",
    "    \"\"\"\n",
    "    genotype = \"C C \tC T \tC T \tC T \tT T\"\n",
    "    Allele frequency:\n",
    "    count the \"Cs\", and the \"Ts\"\n",
    "    f_C = # of Cs\n",
    "    f_T = # of T's\n",
    "    \"\"\"\n",
    "\n",
    "    \n",
    "    return np.array([genotype.str.count(arr[0]).sum(), genotype.str.count(arr[1]).sum()])\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Hardy Weinberg Equilibrium"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Function takes as input the minor allele frequency (p) and the population genotype counts\n",
    "## returns the p-value from the chisq test\n",
    "import scipy\n",
    "def HWEChiSq (p, populationGenotypeCounts):\n",
    "    \"\"\"\n",
    "        |  CC | CT  | TT |\n",
    "    BP  | 270 | 957 | 771|\n",
    "    compute the HWE\n",
    "      -- p = frequency of the C allele\n",
    "      -- q = 1-p = frequency of the T allele\n",
    "      -- if the population genotype is in Hardy Weinberg Equilibrium, we would expect the genotype frequencies to be\n",
    "          CC = p^2*N \n",
    "          CT = 2pq*N \n",
    "          TT = q^2*N \n",
    "      -- to compute the deviation from HWE\n",
    "         (observed - expected)^2 / expected\n",
    "      -- do a chi.squared test to check if the deviation is significant\n",
    "    \"\"\"\n",
    "    total = sum(populationGenotypeCounts)\n",
    "\n",
    "    q = 1-p\n",
    "    \n",
    "#     observed = populationGenotypeCounts[0]\n",
    "    CC_expected = p**2*total\n",
    "    \n",
    "#     CC = (observed-expected)**2/expected\n",
    "    \n",
    "    \n",
    "#     observed = populationGenotypeCounts[1]\n",
    "    CT_expected = 2*p*q*total\n",
    "    \n",
    "#     CT = (observed-expected)**2/expected\n",
    "    \n",
    "#     observed = populationGenotypeCounts[2]\n",
    "    TT_expected = q**2*total\n",
    "    \n",
    "#     TT = (observed-expected)**2/expected\n",
    "    \n",
    "   \n",
    "#     from scipy.stats.distributions import chi2\n",
    "#     return chi2.sf(CC+CT+TT, 1)\n",
    "    test_result = scipy.stats.chisquare(f_obs=populationGenotypeCounts, f_exp=np.array([CC_expected, CT_expected, TT_expected]), ddof=1, axis=0)\n",
    "    return test_result[1]\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Compute the Odds Ratio\n",
    "## takes as input a 2X2  confusion matrix\n",
    "## returns the odds ratio\n",
    "def computeOR (confusionMatrix):\n",
    "    ad = confusionMatrix[0][0] * confusionMatrix[1][1]\n",
    "    bc = confusionMatrix[1][0] * confusionMatrix[0][1]\n",
    "    return ad/bc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Execute a chisq test to determine if the odds ratio is significant\n",
    "## return the p-value from the chisq test\n",
    "from scipy.stats import chisquare\n",
    "def getPValue (confusionMatrix):\n",
    "    return chisquare(confusionMatrix).pvalue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "## plot the data\n",
    "## manhattan plot\n",
    "## the X axis represents the chromosomes \n",
    "## the y-axis is the -log(p.value)\n",
    "def manhattanPlot (diseaseResults):\n",
    "    \"\"\"\n",
    "    group data by chromosomes, SNP before plotting\n",
    "    \"\"\"\n",
    "    return "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Write code to run the GWAS\n",
    "\n",
    "## Write a function that will take the following inputs:\n",
    "## (a) control files - 58C and NBS\n",
    "## (b) the disease - T2D, T1D, HT, BD, CD, RA\n",
    "## (c) the chromosome number\n",
    "\n",
    "## Your function should output a csv file with the following information\n",
    "## SNP Id, RSID, CHROMOSOME NUMBER, MINOR ALLELE, MAJOR ALLELE, MINOR ALLELE FREQUENCY, MAJOR ALLELE FREQUENCY, \n",
    "## ODDS RATIO (MINOR vs MAJOR ALLELE), P-Value for ODDS RATIO, Chi square for the ODDS RATIO, HWE deviation P-VALUE\n",
    "## \n",
    "from scipy.stats import chi2_contingency\n",
    "\n",
    "def gwas (disease, chrom_num, row, t1d, control_58c_body, combined_t1d):\n",
    "    \"\"\"\n",
    "    1. Load the control file for the specified chromosome number. Since there are two control populations, \n",
    "       it would be best to combine the two controls together\n",
    "    2. Load the disease file for the given chromosome number\n",
    "    3. Split the header from the SNP data for the controls and the disease\n",
    "    3. Load the SNP file -- this gives you the mapping between the WTCCC SNP ids and the RSID\n",
    "    4. For each SNP in the disease file and its matching SNP in the control file:\n",
    "       -- to make it easier, before you begin do a check for zygosity. If either the controls or the disease \n",
    "       are homozygous, skip the SNP\n",
    "       genotype CC, TT : skip\n",
    "       genotype CC CT TT : run the following analysis\n",
    "       a. get the minor and major allele counts and compute the minor allele frequency for the disease population\n",
    "       b. get the minor and major allele counts and compute the minor allele frequency for the control population\n",
    "       c. conduct the allelic test -- compute the odds ratios to test the frequency of the minor allele \n",
    "          in the disease population compared to the control population\n",
    "          -- you can build a confusion matrix based on the counts that you have calculated in a and b\n",
    "          -- compute your odds ratios based on the minor allele of the control population\n",
    "          \n",
    "          |                  | C    | T    | \n",
    "          |------------------+ ---  + ---  +\n",
    "          | Bipolar Disorder | 1529 | 2469 |\n",
    "          |------------------+----  + ---  +\n",
    "          | Healthy Controls | 2270 | 3738 |\n",
    "          |------------------+----- +----- +\n",
    "          \n",
    "          (Test : your OR for above should be :1.019)\n",
    "       d. Do a chi square test to test the significant and record the p-value\n",
    "       d. for each SNP calculate the deviation from Hardy-Weinberg equilibrium\n",
    "       e. create an output row containing:\n",
    "       SNP Id, RSID, CHROMOSOME NUMBER, MINOR ALLELE, MAJOR ALLELE, MINOR ALLELE FREQUENCY, \n",
    "       MAJOR ALLELE FREQUENCY, ODDS RATIO (MINOR vs MAJOR ALLELE), \n",
    "       P-Value for ODDS RATIO, CHI_SQ for the ODDS RATIO, HWE deviation P-VALUE  \n",
    "    5. Save the result file for the chromosome\n",
    "    \"\"\"\n",
    "    \n",
    "#     dataName = \"LittleBoy\"\n",
    "#     path = \"\"\n",
    "#     if(dataName=='Japan'):\n",
    "#         path = \"/Users/varun/Desktop/BeaverWorks/Week3Public/BWSI_100set/\"\n",
    "#     elif(dataName=='LittleBoy'):\n",
    "#         path = \"/Users/varun/Desktop/BeaverWorks/Week3Public/BWSI_100set/LittleBoy\"\n",
    "#     elif(dataName=='FatMan'):\n",
    "#         path = \"/Users/varun/Desktop/BeaverWorks/Week3Public/BWSI_100set/FatMan\"\n",
    "        \n",
    "    \n",
    "#     if(chrom_num<10):\n",
    "#         fileName = path + '/' + str(disease) + '/Affx_gt_' + str(disease) + '_Chiamo_0' + str(chrom_num) + '.tped.gz'   \n",
    "#     else:\n",
    "#         fileName = path + '/' + str(disease) + '/Affx_gt_' + str(disease) + '_Chiamo_' + str(chrom_num) + '.tped.gz'\n",
    "#     data = pd.read_csv(fileName, compression='gzip', header=None, sep='\\t')\n",
    "\n",
    "#     if chrom_num < 10:\n",
    "#         chrom_num = \"0\"+str(chrom_num)\n",
    "#     control_file = path + '/58C/Affx_gt_58C_Chiamo_' + str(chrom_num) + '.tped.gz'\n",
    "#     control_58c = pd.read_csv(control_file, compression='gzip', header=None, sep='\\t')\n",
    "\n",
    "\n",
    "#     control_58c_header = control_58c.loc[:, :3]\n",
    "#     control_58c_body = control_58c.loc[:, 4:]\n",
    "\n",
    "#     ## Load the test data \n",
    "#     #     t1d_22_file = '/Users/varun/Desktop/BeaverWorks/Week3Public/BWSI_100set/T1D/Affx_gt_T2D_Chiamo_22.tped.gz'\n",
    "#     t1d = data\n",
    "\n",
    "#     t1d_header = data.loc[:, :3]\n",
    "#     t1d_body = data.loc[:,4:]\n",
    "\n",
    "#     snp_file = path + '/58C/snps/snps_'+str(chrom_num)\n",
    "#     snp = pd.read_csv(snp_file, header=None, sep='\\t')\n",
    "\n",
    "\n",
    "#     combined = pd.concat([control_58c_header, snp], axis=1)\n",
    "    t1d_1 = t1d.loc[row, 4:]\n",
    "    control_58c_1 = control_58c_body.loc[row]\n",
    "    control_58c_string = control_58c_1.to_string()\n",
    "    t1d_string = t1d_1.to_string()\n",
    "    if(findAlleles(t1d_1)!=None and findAlleles(control_58c_1)!=None):\n",
    "        snp = int(str(t1d.loc[row, 1])[6:])\n",
    "        rsid=int(str(combined_t1d.loc[row, 4])[2:])\n",
    "        minorAllele = findAlleles(t1d_1)[0]\n",
    "        majorAllele = findAlleles(t1d_1)[1]\n",
    "        t1_minor = computeAlleleFrequency(getAlleleCounts(t1d_1, findAlleles(t1d_1))[0], getAlleleCounts(t1d_1, findAlleles(t1d_1))[1])\n",
    "        if(t1_minor < 0.01):\n",
    "            return pd.Series([])\n",
    "        t1_major = 1 - t1_minor\n",
    "        control_minor = computeAlleleFrequency(getAlleleCounts(control_58c_1, findAlleles(control_58c_1))[0], getAlleleCounts(control_58c_1, findAlleles(control_58c_1))[1])\n",
    "        if(control_minor < 0.01):\n",
    "            return pd.Series([])\n",
    "        HWEControl=HWEChiSq(control_minor, [control_58c_string.count(minorAllele + ' ' + minorAllele), control_58c_string.count(minorAllele + ' ' + majorAllele), control_58c_string.count(majorAllele + ' ' + majorAllele)])\n",
    "        if(HWEControl < 0.05):\n",
    "            return pd.Series([])\n",
    "        control_major = 1-control_minor\n",
    "        oddsRatio=computeOR([[t1_minor, t1_major], [control_minor, control_major]])\n",
    "        oddsRatioP=chi2_contingency(np.array([[t1d_string.count(minorAllele),t1d_string.count(majorAllele)], [control_58c_string.count(minorAllele),control_58c_string.count(majorAllele)]]))[1]\n",
    "        HWEDisease=HWEChiSq(t1_minor, [t1d_string.count(minorAllele + ' ' + minorAllele), t1d_string.count(minorAllele + ' ' + majorAllele), t1d_string.count(majorAllele + ' ' + majorAllele)])\n",
    "        logOddsRatioP = np.log10(oddsRatioP)*-1\n",
    "        logHWEControl= np.log10(HWEControl)*-1\n",
    "        return pd.Series([chrom_num, snp, rsid, minorAllele, majorAllele, control_minor, control_major, t1_minor, t1_major, oddsRatio, oddsRatioP, HWEControl, logOddsRatioP, logHWEControl])\n",
    "    else:\n",
    "        return pd.Series([])\n",
    "# arr = gwas('T1D', 22)\n",
    "# # print(len(arr[0])\n",
    "# df = pd.DataFrame({\"Chromosome\":arr[0],\n",
    "#                   \"SNP\": arr[1],\n",
    "#                    \"Minor Allele\": arr[9],\n",
    "#                    \"Major Allele\": arr[10],\n",
    "#                   \"Minor Allele Control Frequency\": arr[2],\n",
    "#                   \"Major Allele Control Frequency\": arr[3],\n",
    "#                   \"Minor Allele Disease Frequency\": arr[4],\n",
    "#                   \"Major Allele Disease Frequency\": arr[5],\n",
    "#                   \"Odds Ratio\": arr[6],\n",
    "#                   \"Odds Ratio p-value\": arr[7],\n",
    "#                   \"HWE\": arr[8]})\n",
    "# df.to_csv(\"data.csv\", sep='\\t')\n",
    "# print(\"Done!\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "# print(len(arr[1]))\n",
    "# print(len(arr[8]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/javascript": [
       "IPython.OutputArea.prototype._should_scroll = function(lines) {\n",
       "    return false;\n",
       "}"
      ],
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%javascript\n",
    "IPython.OutputArea.prototype._should_scroll = function(lines) {\n",
    "    return false;\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chromosome 1\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 0\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 1\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 2\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 3\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 4\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 5\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 6\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 7\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 8\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 9\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 10\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 11\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 12\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 13\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 14\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 15\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 16\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 17\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 18\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 19\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 20\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 21\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 22\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 23\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 24\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 25\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 26\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 27\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 28\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 29\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 30\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 31\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 32\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 33\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 34\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 35\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 36\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 37\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 38\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 39\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 40\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 41\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 42\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 43\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 44\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 45\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 46\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 47\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 48\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 49\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 50\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 51\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 52\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 53\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 54\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 55\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 56\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 57\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 58\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 59\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 60\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 61\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 62\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 63\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 64\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 65\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 66\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 67\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 68\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 69\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 70\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 71\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 72\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 73\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 74\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 75\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 76\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 77\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 78\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 79\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 80\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 81\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 82\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 83\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 84\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 85\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 86\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 87\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 88\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 89\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 90\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 91\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 92\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 93\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 94\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 95\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 96\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 97\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 98\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 99\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 100\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 101\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 102\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 103\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 104\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 105\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 106\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 107\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 108\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 109\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 110\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 111\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 112\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 113\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 114\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 115\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 116\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 117\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 118\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 119\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 120\n",
      "(1000, 2004)\n",
      "(1000, 1504)\n",
      "SNP 121\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-53-799b03d31fab>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     81\u001b[0m     \u001b[0mdata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"data_long.csv\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msep\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'\\t'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     82\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 83\u001b[0;31m \u001b[0mrunGWAS\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'T1D'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     84\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Done!\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-53-799b03d31fab>\u001b[0m in \u001b[0;36mrunGWAS\u001b[0;34m(disease)\u001b[0m\n\u001b[1;32m     60\u001b[0m             \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"SNP \"\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mj\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     61\u001b[0m             \u001b[0;32mif\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mempty\u001b[0m\u001b[0;34m!=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 62\u001b[0;31m                 \u001b[0mdata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     63\u001b[0m                 \u001b[0mindex\u001b[0m \u001b[0;34m+=\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     64\u001b[0m         \u001b[0mdata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"data_long.csv\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msep\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'\\t'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m__setitem__\u001b[0;34m(self, key, value)\u001b[0m\n\u001b[1;32m    188\u001b[0m             \u001b[0mkey\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcom\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply_if_callable\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    189\u001b[0m         \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_setitem_indexer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 190\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_setitem_with_indexer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    191\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    192\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_validate_key\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m_setitem_with_indexer\u001b[0;34m(self, indexer, value)\u001b[0m\n\u001b[1;32m    613\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    614\u001b[0m                     \u001b[0;32mfor\u001b[0m \u001b[0mitem\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mv\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mzip\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlabels\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 615\u001b[0;31m                         \u001b[0msetter\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mitem\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mv\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    616\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    617\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36msetter\u001b[0;34m(item, v)\u001b[0m\n\u001b[1;32m    541\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    542\u001b[0m                 \u001b[0;31m# reset the sliced object if unique\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 543\u001b[0;31m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mitem\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0ms\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    544\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    545\u001b[0m             \u001b[0;32mdef\u001b[0m \u001b[0mcan_do_equal_len\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m__setitem__\u001b[0;34m(self, key, value)\u001b[0m\n\u001b[1;32m   3368\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3369\u001b[0m             \u001b[0;31m# set column\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3370\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_set_item\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3371\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3372\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_setitem_slice\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m_set_item\u001b[0;34m(self, key, value)\u001b[0m\n\u001b[1;32m   3444\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_ensure_valid_index\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3445\u001b[0m         \u001b[0mvalue\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_sanitize_column\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3446\u001b[0;31m         \u001b[0mNDFrame\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_set_item\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3447\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3448\u001b[0m         \u001b[0;31m# check if we are modifying a copy\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m_set_item\u001b[0;34m(self, key, value)\u001b[0m\n\u001b[1;32m   3170\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3171\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_set_item\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3172\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_data\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3173\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_clear_item_cache\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3174\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pandas/core/internals/managers.py\u001b[0m in \u001b[0;36mset\u001b[0;34m(self, item, value)\u001b[0m\n\u001b[1;32m   1027\u001b[0m         \u001b[0;31m# TODO(EA): Remove an is_extension_ when all extension types satisfy\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1028\u001b[0m         \u001b[0;31m# the interface\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1029\u001b[0;31m         value_is_extension_type = (is_extension_type(value) or\n\u001b[0m\u001b[1;32m   1030\u001b[0m                                    is_extension_array_dtype(value))\n\u001b[1;32m   1031\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pandas/core/dtypes/common.py\u001b[0m in \u001b[0;36mis_extension_type\u001b[0;34m(arr)\u001b[0m\n\u001b[1;32m   1691\u001b[0m     \"\"\"\n\u001b[1;32m   1692\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1693\u001b[0;31m     \u001b[0;32mif\u001b[0m \u001b[0mis_categorical\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1694\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1695\u001b[0m     \u001b[0;32melif\u001b[0m \u001b[0mis_sparse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pandas/core/dtypes/common.py\u001b[0m in \u001b[0;36mis_categorical\u001b[0;34m(arr)\u001b[0m\n\u001b[1;32m    289\u001b[0m     \"\"\"\n\u001b[1;32m    290\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 291\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mABCCategorical\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0mis_categorical_dtype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    292\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    293\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pandas/core/dtypes/common.py\u001b[0m in \u001b[0;36mis_categorical_dtype\u001b[0;34m(arr_or_dtype)\u001b[0m\n\u001b[1;32m    600\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0marr_or_dtype\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    601\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 602\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0mCategoricalDtype\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mis_dtype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marr_or_dtype\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    603\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    604\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pandas/core/dtypes/base.py\u001b[0m in \u001b[0;36mis_dtype\u001b[0;34m(cls, dtype)\u001b[0m\n\u001b[1;32m     99\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    100\u001b[0m         if isinstance(dtype, (ABCSeries, ABCIndexClass,\n\u001b[0;32m--> 101\u001b[0;31m                               ABCDataFrame, np.dtype)):\n\u001b[0m\u001b[1;32m    102\u001b[0m             \u001b[0;31m# https://github.com/pandas-dev/pandas/issues/22960\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    103\u001b[0m             \u001b[0;31m# avoid passing data to `construct_from_string`. This could\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "def runGWAS (disease):\n",
    "    \"\"\"\n",
    "    1. For each chromosome 1:22 call the gwas function\n",
    "    2. Merge all the result files into one \n",
    "    3. Filter your data\n",
    "       a. remove all SNPs that deviate signficantly from HWE equilibrium -- these may indicate an error or \n",
    "          population specific deviation. For this remove all SNPs for which HWE deviation P-Value < 0.05\n",
    "       b. remove all SNPs that have minor allele frequency less than 1% -- not enough data\n",
    "    4. Draw a manhattan plot\n",
    "       c. draw a line at your Bonferroni threshold (Bonferroni correction): \n",
    "       if your significance is set at p.value < 0.05, then Bonferroni correction = 0.05/number of tests that you ran\n",
    "    \"\"\"\n",
    "    import os\n",
    "    \n",
    "    dataName = \"LittleBoy\"\n",
    "    path = \"\"\n",
    "    if(dataName=='Japan'):\n",
    "        path = \"/Users/varun/Desktop/BeaverWorks/Week3Public/BWSI_100set/\"\n",
    "    elif(dataName=='LittleBoy'):\n",
    "        path = \"/Users/varun/Desktop/BeaverWorks/Week3Public/BWSI_100set/LittleBoy\"\n",
    "    elif(dataName=='FatMan'):\n",
    "        path = \"/Users/varun/Desktop/BeaverWorks/Week3Public/BWSI_100set/FatMan\"\n",
    "    \n",
    "    data = pd.DataFrame(columns = [\"Chromosome\",\"SNP\",\"RSID\", \"Minor Allele\",\"Major Allele\",\"Minor Allele Control Frequency\",\"Major Allele Control Frequency\",\"Minor Allele Disease Frequency\",\"Major Allele Disease Frequency\",\"Odds Ratio\",\"Odds Ratio p-value\",\"HWE\", \"-log(Odds Ratio p-value)\", \"-log(HWE)\"])\n",
    "    index = 0\n",
    "    for chrom_num in range(1, 23):\n",
    "        print(\"Chromosome \" + str(chrom_num))\n",
    "        \n",
    "        if(chrom_num<10):\n",
    "            fileName = path + '/' + str(disease) + '/Affx_gt_' + str(disease) + '_Chiamo_0' + str(chrom_num) + '.tped.gz'   \n",
    "        else:\n",
    "            fileName = path + '/' + str(disease) + '/Affx_gt_' + str(disease) + '_Chiamo_' + str(chrom_num) + '.tped.gz'\n",
    "        data = pd.read_csv(fileName, compression='gzip', header=None, sep='\\t')\n",
    "        \n",
    "        if chrom_num < 10:\n",
    "            chrom_num = \"0\"+str(chrom_num)\n",
    "        control_file = path + '/58C/Affx_gt_58C_Chiamo_' + str(chrom_num) + '.tped.gz'\n",
    "        control_58c = pd.read_csv(control_file, compression='gzip', header=None, sep='\\t')\n",
    "        \n",
    "        \n",
    "        control_58c_header = control_58c.loc[:, :3]\n",
    "        control_58c_body = control_58c.loc[:, 4:]\n",
    "\n",
    "        ## Load the test data \n",
    "        #     t1d_22_file = '/Users/varun/Desktop/BeaverWorks/Week3Public/BWSI_100set/T1D/Affx_gt_T2D_Chiamo_22.tped.gz'\n",
    "        t1d = data\n",
    "\n",
    "        t1d_header = data.loc[:, :3]\n",
    "        t1d_body = data.loc[:,4:]\n",
    "\n",
    "        snp_file = path + '/58C/snps/snps_'+str(chrom_num)\n",
    "        snp = pd.read_csv(snp_file, header=None, sep='\\t')\n",
    "        \n",
    "    \n",
    "        \n",
    "        \n",
    "        combined = pd.concat([control_58c_header, snp], axis=1)\n",
    "        for j in range(1000):\n",
    "            x = gwas(disease, chrom_num, j, t1d, control_58c_body, combined)\n",
    "            print(\"SNP \" + str(j))\n",
    "            if(x.empty!=True):\n",
    "                data.loc[index] = x\n",
    "                index += 1\n",
    "        data.to_csv(\"data_long.csv\", sep='\\t')\n",
    "            \n",
    "\n",
    "    \n",
    "    \n",
    "#     df = pd.DataFrame({\"Chromosome\":arr[0],\n",
    "#                   \"SNP\": arr[1],\n",
    "#                    \"Minor Allele\": arr[9],\n",
    "#                    \"Major Allele\": arr[10],\n",
    "#                   \"Minor Allele Control Frequency\": arr[2],\n",
    "#                   \"Major Allele Control Frequency\": arr[3],\n",
    "#                   \"Minor Allele Disease Frequency\": arr[4],\n",
    "#                   \"Major Allele Disease Frequency\": arr[5],\n",
    "#                   \"Odds Ratio\": arr[6],\n",
    "#                   \"Odds Ratio p-value\": arr[7],\n",
    "#                   \"HWE\": arr[8]})\n",
    "    \n",
    "    data.to_csv(\"data_long.csv\", sep='\\t')\n",
    "\n",
    "runGWAS('T1D')\n",
    "print(\"Done!\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src=\"manhattan.png\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## How many SNPs did you identify that exceeded the Bonferroni-level of significance?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## How much can you trust your results\n",
    "\n",
    "Typically, you want to examine the genome-wide distribution of your chi_square distribution against an expected null distribution. Visually you can do this using a quantile-quantile plot. Significant deviation from the null distribution may point to population stratification, familial relationships, technical bias, poor sample collection or sometimes even undetected sample duplications.\n",
    "(Slide courtesy - Dr. Chirag Patel)\n",
    "\n",
    "<img src=\"qqplot_example.png\">\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Based on your results do a quantile-quantile plot for your data\n",
    "\n",
    "##load your results \n",
    "##sort/order the -log10 of your p values in increasing order\n",
    "##generate an expected set of -log10(p_values)\n",
    "##plot the expected values on the x-axis and the observed -log10 (pvalues) on the y axis\n",
    "##Alternatively use: p = observed p values\n",
    "##stats.probplot(p, dist=\"norm\", plot=pylab)\n",
    "##pylab.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Genomic Inflation Factor\n",
    "\n",
    "Mathematically, you can correct for some of the deviation by computing an inflation factor and then correcting for this inflation factor. This quantity is termed as genomic inflation factor ($\\lambda$)\n",
    "\n",
    "$\\lambda$ = observed median of test statistic distribution / expected median of the test statistic \n",
    "distribution.  For 1 degree of freedom, the X2 distribution has an expected median of 0.455\n",
    "\n",
    "$\\lambda$ <= 1.05 is considered acceptable. >1.1 is troubling, and indicates there is some inflation of the p values.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Compute the genomic inflation factor\n",
    "\n",
    "##labmda =  median(CHI2)/qchisq(0.5, 1)\n",
    "##if (gif > 1.05), correct your computed Chi.square values and recompute the p_values\n",
    "##chi2_corrected = CHI2/lambda, \n",
    "##pval_corrected = pchisq(chi2_corrected, 1, lower.tail = F)\n",
    "##re-draw the qqplot using the corrected pvalues (pval_corrected)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## What does this mean?\n",
    "\n",
    "Identify a SNP that is significant (above the Bonferroni threshold). \n",
    "What is the p-value of association and odds ratio for the SNP?\n",
    "What is the interpretation of the odds ratio? \n",
    "What gene is this SNP associated with and its putative function in your disease pathogenesis? \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Polygenic Risk Scores (PRS)\n",
    "\n",
    "A PRS is the cumulative risk of all SNP for a single patient. For example, if a patient has a 2 snps with OR=2 each, then their PRS is 4. The underlying assumption here is that each allele has an equal contribution and there is no interation between SNPS, that is the SNPS are independent. \n",
    "\n",
    "$prs =\\sum_{i=1}^{m}-log_{10}(x_i) \\cdot n_i$"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
